<body>
<p>Primary package for sequence alignment.</p>

<h2>Sequence Alignment</h2>

<h3>What is an &quot;Edit Distance&quot; operation?</h3>

<p>An edit distance is the number of operations required to transform one string into another. e.g. You can transform <code>cat</code> into <code>mat</code>
by the operation <code>C -&gt; M</code>. The final result referring to the <code>alignment</code> is a combination of the alignment score (in this case 1 operation) and the actions (change a C into a M).
</p>

<h3>So what is sequence alignment?</h3>

<p>Sequence alignment is the generalization of the edit distance operation by extensions to sequences. Sequence alignment determines the operations
required to transform a DNA string (for example) into another. e.g. Transform <code>ACATG</code> into <code>ACG</code> would be accomplished by inserting <code>AT</code> after the <code>C</code>.

<h3>Edit distance Actions</h3>

<p>The actions and associated penalties for sequence alignment are:</p>
<ul>
<li> SAME, cost = 0
<li> The beginning of a INSERT/DELETION operation, cost = 2
<li> Continuing an INSERT/DELETION operation, cost = 1
<li> Changing one char into another, cost = 1
<li> Cost of matching against an N, cost = 0
<li> Cost of matching against an N when overhanging a reference template, cost = 1
</ul>

<p>There are special/more advanced cases for Protein and Complete Genomics read formats.</p>

<h3>What are the attributes of sequence alignment that are unique?</h3>

<p>
Unlike traditional edit distance operations, we don't want the list of operations that convert a read into a huge reference template/chromosome.
</p>
<p>
Here we want the minimal number of operations that full describes the read. Effectively the read is mapped as a error tolerant sub-string of a reference sequence.
</p>
<p>
Other differences include that a read may overhang ends of a template, the costs of the Ns change (see above).
</p>
<p>
The exact position where the read is expected to be placed on the reference (where matching will take place) 
is not always known exactly. Some of the edit distance functions allow for a loose interpretation of the start
position and can adjust it if they find a better start position.
</p> 
<p>
Sequences need be matched in the forward direction (<code>ATGCGA</code>) but also need to be matched in the reverse complement
(RC) direction. Reverse complementation means to reverse the string and switch A/T and C/G. This example string would be
reversed to be <code>AGCGTA</code> and then complemented to be <code>TCGCAT</code>.
</p>

<h2>Real Time Aligner architecture</h2>

<p>The Real Time Aligner is a suite of classes that are used as components of the sequence alignment solution.</p>

<h3>Base classes</h3>

<p>The two base classes are the <code>EditDistance</code> and <code>UnidirectionalEditDistance</code> classes.
</p>

<p>The fundamental difference between the two is that the <code>EditDistance</code> allows the user to specify whether the match be made in the forward or reverse complement 
direction.</p> 

<p>The <code>UnidirectionalEditDistance</code> class does not know about the reverse complement process.</p>

<h3>Wrapping the reverse complement process</h3>

<p><code>RcEditDistance</code> is a wrapping class that implements <code>EditDistance</code>, but internally looks after the
reverse complementing process by storing a RC version of the template and using the forward or RC versions of the reference
as required.</p>

<p>As the seeded aligner and others may cache a template index, the <code>RcEditDistance</code> constructor can take two
edit distance classes, one for the forward and one for the RC direction.
</p>
 
<h3>Nesting and pipelining edit distance calls</h3>

<p>
A pipeline of joining edit distance operations together is achieved using the return value from the <code>UnidirectionalEditDistance</code> calls.
</p>
<p>
If an edit distance method doesn't know how to process a sequence it can return <code>NULL</code> to signify further processing later on in the pipeline
should be performed.
<p>
Alternatively, if an edit distance method returns an array of actions these are returned back to the caller. Valid values are either an alignment score (e.g. <code>ATG</code> -&gt; <code>ATG</code>, score=0) or Integer.MAX_VALUE signifying that the alignment score is outside of the maximum value the user deems to be acceptable.
</p>

<h2>Worked example</h2>

<pre>
 EditDistance e = 
     new RcEditDistance(new UnidirectionalPrioritisedEditDistance(new LowerBoundEditDistance(),
                               new <code>MaxOneSubstitutionEditDistance()</code>, 
                               new SeededAligner(), new LoggingOnlyEditDistance(), 
                               new GotohEditDistance(gapOpenPenalty), new MaxIntEditDistance()), 
                        new UnidirectionalPrioritisedEditDistance(new LowerBoundEditDistance(), 
                               new <code>MaxOneSubstitutionEditDistance()</code>, 
                               new SeededAligner(), new LoggingOnlyEditDistance(), 
                               new GotohEditDistance(gapOpenPenalty), new MaxIntEditDistance()));
</pre>


<p>
A typical calling pattern logged from the UnidirectionalPrioritisedEditDistance looks like the following:
<pre>
  [ 1 ] calls= 25210140 nulls= 23256796 MaxInt= 1953344 avg.score= 0 time(s)= 157.731072 com.rtg.alignment.LowerBoundEditDistance
  [ 2 ] calls= 23256796 nulls= 22860312 MaxInt= 0 avg.score= 0 time(s)= 30.062592 <code>com.rtg.alignment.MaxOneSubstitutionEditDistance</code>
  [ 3 ] calls= 22860312 nulls= 143537 MaxInt= 9311280 avg.score= 4 time(s)= 3229.587456 com.rtg.alignment.SeededAligner
  [ 4 ] calls= 143537 nulls= 143537 MaxInt= 0 avg.score= 0 time(s)= 0.169216 com.rtg.alignment.LoggingOnlyEditDistance
  [ 5 ] calls= 143537 nulls= 0 MaxInt= 0 avg.score= 32 time(s)= 97.406208 com.rtg.alignment.GotohEditDistance
  [ 6 ] calls= 0 nulls= 0 MaxInt= 0 avg.score= - time(s)= 0.0 com.rtg.alignment.MaxIntEditDistance
</pre>

<p>The actions above are part of a six-stage pipeline. The order in which the edit distance calls are made are shown by the number
in brackets on the left.</p>

<p>In this example, there were 25 million top level calls to the <code>EditDistance</code> instance. Of the 25 million calls, 23 million calls were returned as NULL. This
means that <code>LowerBoundEditDistance</code> class determined a lower bound score that was less than the threshold that was passed to it. In this example
<code>maxScore</code> was set to 20. </p>

<p>This means that if the alignment score is over 20 then <code>Integer.MAX_VALUE</code> is returned. We can see that this happened 1.9 million times.</p>

<p>The <code>LowerBoundEditDistance</code> is key as it eliminates a large number of bad matches (as comes about by repeat regions). Although the times are sampled, the throughput it approximately
160,000 calls per second (per thread). As these were 100 long reads that is a throughput of 15 million bytes (nucleotides) per second core.
</p>

<p>Of the 23 million NULL values returned by the stage 1 in the queue, 22 million were not processed (returned NULL) because they had more than one
substitution. </p>

<p>Most of the work is performed by the <code>SeededAligner</code> (and it's significantly slower at around 7,000 calls per second/core).</p>
 
<p>For the read/reference regions that don't have decent sized seeds, NULL is returned and read gets passed in the traditional
<code>GotohEditDistance</code> operation. This is the slowest of them all at around 1500 calls per second/core.</p>

<p>The final <code>MaxIntEditDistance</code> is used to housekeeping purposes (and to allow easy insertion of new stages).<p>

<p>The <code>LoggingOnlyEditDistance</code> prints out the first few reads that are passed to it and is for design/testing purposes only.</p>


<h2>LowerBoundEditDistance</h2>

There is some secret sauce in here.

<h2><code>MaxOneSubstitutionEditDistance</code></h2>

<h2>SeededAligner</h2>

There is a lot of secret sauce in here. 

<h2>GotohEditDistance</h2>

</body>
