

<!doctype html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Overview &#8212; RTG Core Operations Manual v3.10</title>
    
    <link rel="stylesheet" href="_static/bizstyle.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '3.10',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/bizstyle.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="RTG Command Reference" href="rtg_command_reference.html" />
    <link rel="prev" title="Table Of Contents" href="index.html" />
    <meta name="viewport" content="width=device-width,initial-scale=1.0">
    <!--[if lt IE 9]>
    <script type="text/javascript" src="_static/css3-mediaqueries.js"></script>
    <![endif]-->
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="rtg_command_reference.html" title="RTG Command Reference"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Table Of Contents"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">RTG Core Operations Manual v3.10</a> &#187;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="index.html">
              <img class="logo" src="_static/html_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Overview</a><ul>
<li><a class="reference internal" href="#introduction">Introduction</a></li>
<li><a class="reference internal" href="#rtg-software-description">RTG software description</a></li>
<li><a class="reference internal" href="#sequence-search-and-alignment">Sequence search and alignment</a><ul>
<li><a class="reference internal" href="#data-formatting">Data formatting</a></li>
<li><a class="reference internal" href="#read-mapping-and-alignment">Read mapping and alignment</a></li>
<li><a class="reference internal" href="#read-mapping-output-files">Read mapping output files</a></li>
<li><a class="reference internal" href="#read-mapping-sensitivity-tuning">Read mapping sensitivity tuning</a></li>
<li><a class="reference internal" href="#protein-search">Protein search</a></li>
<li><a class="reference internal" href="#protein-search-output-files">Protein search output files</a></li>
<li><a class="reference internal" href="#protein-search-sensitivity-tuning">Protein search sensitivity tuning</a></li>
<li><a class="reference internal" href="#benchmarking-and-optimization-utilities">Benchmarking and optimization utilities</a></li>
</ul>
</li>
<li><a class="reference internal" href="#variant-detection-functions">Variant detection functions</a><ul>
<li><a class="reference internal" href="#sequence-variation-snps-indels-and-complex-variants">Sequence variation (SNPs, indels and complex variants)</a></li>
<li><a class="reference internal" href="#sequence-variation-with-mendelian-pedigree">Sequence variation with Mendelian pedigree</a></li>
<li><a class="reference internal" href="#somatic-sequence-variation">Somatic sequence variation</a></li>
<li><a class="reference internal" href="#coverage-analysis">Coverage analysis</a></li>
<li><a class="reference internal" href="#copy-number-variation-cnv-analysis">Copy number variation (CNV) analysis</a></li>
</ul>
</li>
<li><a class="reference internal" href="#standard-input-and-output-file-formats">Standard input and output file formats</a><ul>
<li><a class="reference internal" href="#sam-bam-files-created-by-the-rtg-map-command">SAM/BAM files created by the RTG map command</a></li>
<li><a class="reference internal" href="#variant-caller-output-files">Variant caller output files</a></li>
</ul>
</li>
<li><a class="reference internal" href="#metagenomic-analysis-functions">Metagenomic analysis functions</a><ul>
<li><a class="reference internal" href="#contamination-filtering">Contamination filtering</a></li>
<li><a class="reference internal" href="#taxon-abundance-breakdown">Taxon abundance breakdown</a></li>
<li><a class="reference internal" href="#sample-relationships">Sample relationships</a></li>
<li><a class="reference internal" href="#functional-protein-analysis">Functional protein analysis</a></li>
</ul>
</li>
<li><a class="reference internal" href="#pipelines">Pipelines</a></li>
<li><a class="reference internal" href="#parallel-processing">Parallel processing</a></li>
<li><a class="reference internal" href="#installation-and-deployment">Installation and deployment</a><ul>
<li><a class="reference internal" href="#quick-start-instructions">Quick start instructions</a></li>
<li><a class="reference internal" href="#license-management">License Management</a></li>
</ul>
</li>
<li><a class="reference internal" href="#technical-assistance-and-support">Technical assistance and support</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="index.html"
                        title="previous chapter">Table Of Contents</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="rtg_command_reference.html"
                        title="next chapter">RTG Command Reference</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/overview.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="overview">
<h1>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h1>
<p>This chapter introduces the features, operational options, and
installation requirements of the data analysis software from <a class="reference external" href="http://realtimegenomics.com">Real Time
Genomics</a>.</p>
<div class="section" id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p>RTG software enables the development of fast, efficient software
pipelines for deep genomic analysis. RTG is built on innovative search
technologies and new algorithms designed for processing high volumes of
high-throughput sequencing data from different sequencing technology
platforms. The RTG sequence search and alignment functions enable read
mapping and protein searches with a unique combination of sensitivity
and speed.</p>
<p>RTG-based data production pipelines support unprecedented breadth and
depth of analysis on genomic data, transforming researcher visibility
into DNA sequence analysis and biological investigation. A comprehensive
suite of easy-to-integrate data analysis functions increases the
productivity of bioinformatics specialists, freeing them to develop
analytical solutions that amplify the investigative ability unique to
their organization.</p>
<p>RTG software supports a variety of research and medical genomics
applications, such as:</p>
<ul class="simple">
<li><strong>Medical Genomic Research</strong> &#8211; Compare sequence variants and
structural variation between normal and disease genomes, or over a
disease progression in the same individual to identity causal loci.</li>
<li><strong>Personalized Medicine</strong> &#8211; Establish reliable, high-throughput
processing pipelines that analyze individual human genomes compared
to one or more reference genomes. Use RTG software for detection of
sequence variants (SNP and indel calling, intersection scripting), as
well as structural variation (coverage depth, and copy number
variation).</li>
<li><strong>Model Organisms and Basic Research</strong> &#8211; Utilize RTG mapping and
variant detection commands for focused research applications such as
metagenomic species identification and frequency, and metabolic
pathway analysis. Map microbial communities to generate gapped
alignments of both DNA and protein sequence data.</li>
<li><strong>Plant Genomics</strong> &#8211; Enable investigations of new crop species and
variant detection in genetically diverse strains by leveraging RTG&#8217;s
highly sensitive sequence search capabilities for strain and
cross-species mapping applications. Flexible sensitivity tuning
controls allow investigators to accommodate very high error rates
associated with unique combinations of sequencing system error,
genome-specific mutation, and aggressive cross-species comparisons.</li>
</ul>
</div>
<div class="section" id="rtg-software-description">
<h2>RTG software description<a class="headerlink" href="#rtg-software-description" title="Permalink to this headline">¶</a></h2>
<p>RTG software is delivered as a single executable with multiple commands
executed through a command line interface (CLI). Commands are delivered
in product packages, and for commercial users each command can be
independently enabled through a license key.</p>
<p>Usage:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>rtg COMMAND [OPTIONS] &lt;REQUIRED&gt;
</pre></div>
</div>
<p>RTG software delivers features in four areas:</p>
<ul class="simple">
<li><strong>Sequence Search and Alignment</strong> &#8211; RTG software uses patented
sequence search technology for the rapid production of genomic
sequence data. The <code class="docutils literal"><span class="pre">map</span></code> command implements read mapping and gapped
alignment of sequence data against a reference. The <code class="docutils literal"><span class="pre">mapx</span></code> command
searches translated sequence data against a protein database.</li>
<li><strong>Data Analysis</strong> &#8211; RTG software supports two pipelines for data
analysis - variant detection and metagenomics. Purpose-built variant
detection pipeline functions include several commands to identify
small sequence variants, a <code class="docutils literal"><span class="pre">cnv</span></code> command to report copy number
variation statistics for structural variation, and a <code class="docutils literal"><span class="pre">coverage</span></code>
command to report read depth across a reference.</li>
<li><strong>Reporting Options</strong> &#8211; Standard result formats and utility commands
report results for validation, and ease development of custom scripts
for analysis. Scripts that produce publication quality graphics for
visualization of data analysis results are available through Real
Time Genomics technical support.</li>
<li><strong>Data Center Deployment</strong> &#8211; RTG software supports typical data
center standards for enterprise deployment. RTG provides automated
installation and supports industry standard operating environments
and data processing systems to help maintain total cost of ownership
objectives in enterprise data centers. The RTG software can be run in
compute clusters of varying sizes, and commands take advantage of
multi-core processors by default.</li>
</ul>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last">For detailed information about RTG command syntax and
usage of individual commands, refer to <a class="reference internal" href="rtg_command_reference.html#rtg-command-reference"><span class="std std-ref">RTG Command Reference</span></a>.</p>
</div>
</div>
<div class="section" id="sequence-search-and-alignment">
<h2>Sequence search and alignment<a class="headerlink" href="#sequence-search-and-alignment" title="Permalink to this headline">¶</a></h2>
<p>RTG software uses an edit-distance alignment score to determine best fit
and alignment accuracy.</p>
<p>RTG software includes optimal sensitivity settings for error and
mutation rates, plus command line controls and simulation tools that
allow investigators to calibrate sensitivity settings for specific data
sets. Extensive filtering and reporting options allow complete control
over reported alignments, which leads to greater flexibility for
downstream analysis functions.</p>
<p>Key functionality of RTG sequence search and alignment includes:</p>
<ul class="simple">
<li>Read mapping by nucleotide sequence alignment to a reference genome</li>
<li>Protein database searching by translated nucleotide sequence searches
against protein databases</li>
<li>Sensitivity tuning using parameter options for substitutions, indels,
indel lengths, word or step sizes, and alignment scores</li>
<li>Filtering and reporting ambiguous reads that map to multiple
locations</li>
<li>Benchmarking and optimization using simulation and evaluation
commands</li>
</ul>
<p>RTG mapping commands have the following characteristics:</p>
<ul class="simple">
<li>Eliminates need for genome indexing</li>
<li>Aligns sequence reads of any length</li>
<li>Allows high mismatch levels for increased sensitivity in longer reads</li>
<li>Allows detection of short indels with single end (SE) or paired end
(PE) data</li>
<li>Can optionally guarantee the mapping of reads with at least a
specified number of substitutions and indels</li>
<li>Supports a wide range of alignment scores</li>
</ul>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p>For detailed information about sequence search and
alignment functionality, refer to <a class="reference internal" href="rtg_command_reference.html#map"><span class="std std-ref">Command Reference, map</span></a>.</p>
<p class="last">For more information about the RTG integrated software pipeline, refer to <a class="reference internal" href="product_usage.html#baseline-product-usage"><span class="std std-ref">RTG product usage - baseline progressions</span></a></p>
</div>
<div class="section" id="data-formatting">
<h3>Data formatting<a class="headerlink" href="#data-formatting" title="Permalink to this headline">¶</a></h3>
<p>Prior to RTG data production, reference genome and sometimes read data
sequence files are typically first converted to the RTG Sequence Data
File (SDF) format. This is an efficient storage format optimized for
fast retrieval during data processing.</p>
<p>The RTG <code class="docutils literal"><span class="pre">format</span></code> / <code class="docutils literal"><span class="pre">cg2sdf</span></code> commands converts sequencing system read and
reference genome sequence data into the SDF format. The <code class="docutils literal"><span class="pre">format</span></code> command
accepts source data in standard file formats (such as FASTA / FASTQ /
SAM / BAM) and maintains the integrity and consistency of the source
data during the conversion to SDF. Similarly, the <code class="docutils literal"><span class="pre">cg2sdf</span></code> command
accepts data in the custom data format used for read data by Complete
Genomics, Inc. Read data may be single-end and paired-end reads of fixed
or variable length. Sequence data can be formatted as nucleotide or
protein.</p>
<p>An SDF is a directory containing a series of files that delineate
sequence and quality score information stored in a binary format, along
with metadata that describes the original sequencing system data format
type:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>03/19/2010  12:31 PM    &lt;DIR&gt;          .
03/19/2010  12:31 PM    &lt;DIR&gt;          ..
03/19/2010  12:31 PM             5,038 log
03/19/2010  12:31 PM            24,223 mainIndex
03/19/2010  12:31 PM                75 namedata0
03/19/2010  12:31 PM                 8 nameIndex0
03/19/2010  12:31 PM                56 namepointer0
03/19/2010  12:31 PM        23,267,177 seqdata0
03/19/2010  12:31 PM                56 seqpointer0
03/19/2010  12:31 PM                 8 sequenceIndex0
             8 File(s)      23,296,641 bytes
             2 Dir(s)  400,984,870,912 bytes free
</pre></div>
</div>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last">For detailed information about formatting sequencing
system reads to RTG SDF, refer to <a class="reference internal" href="rtg_command_reference.html#data-formatting-commands"><span class="std std-ref">Data Formatting Commands</span></a></p>
</div>
</div>
<div class="section" id="read-mapping-and-alignment">
<h3>Read mapping and alignment<a class="headerlink" href="#read-mapping-and-alignment" title="Permalink to this headline">¶</a></h3>
<p>The <code class="docutils literal"><span class="pre">map</span></code> command implements read mapping and alignment of sequence data
against a reference genome, supporting gapped alignments for both single
and paired-end reads. The <code class="docutils literal"><span class="pre">cgmap</span></code> command performs the same function for
the gapped, paired-end read data from Complete Genomics, Inc.</p>
<p>A summary of the mapping results is displayed at the command line
following execution of the map command, as shown in the paired-end
example below:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>ARM MAPPINGS
    left    right     both
 6650124  6650124 13300248  64.2% mated uniquely (NH = 1)
  186812   186812   373624   1.8% mated ambiguously (NH &gt; 1)
 1538777  1539520  3078297  14.9% unmated uniquely (NH = 1)
   70667    70125   140792   0.7% unmated ambiguously (NH &gt; 1)
       0        0        0   0.0% unmapped due to read frequency (XC = B)
   13624    13946    27570   0.1% unmapped with no matings but too many hits (XC = C)
  109720   109765   219485   1.1% unmapped with poor matings (XC = d)
     984     1003     1987   0.0% unmapped with too many matings (XC = e)
  212158   211688   423846   2.0% unmapped with no matings and poor hits (XC = D)
       0        0        0   0.0% unmapped with no matings and too many good hits (XC = E)
 1569609  1569492  3139101  15.2% unmapped with no hits
10352475 10352475 20704950 100.0% total
</pre></div>
</div>
<p>The following display shows the summary output for single end mapped
data from the map command</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>READ MAPPINGS

 875007  87.5% mapped uniquely (NH = 1)
  25174   2.5% mapped ambiguously (NH &gt; 1)
     71   0.0% unmapped due to read frequency (XC = B)
  88729   8.9% unmapped with too many hits (XC = C)
   8940   0.9% unmapped with poor hits (XC = D)
      0   0.0% unmapped with too many good hits (XC = E)
   2079   0.2% unmapped with no hits
1000000 100.0% total
</pre></div>
</div>
<p>Read mapping commands also produce HTML summary reports containing more
information about mapping results.</p>
</div>
<div class="section" id="read-mapping-output-files">
<h3>Read mapping output files<a class="headerlink" href="#read-mapping-output-files" title="Permalink to this headline">¶</a></h3>
<p>The map command creates alignment reports in BAM file format and a
summary report file named <code class="docutils literal"><span class="pre">summary.txt</span></code>. There is also a file called
<code class="docutils literal"><span class="pre">progress</span></code> that can be used to monitor overall progress during a run,
and a file named <code class="docutils literal"><span class="pre">map.log</span></code> containing technical information that may be
useful for debugging. Alignment reports may be filtered by alignment
score, and/or unmapped, unmated, and ambiguous reads (those that map to
multiple locations).</p>
<p>When mapping, the output BAM file is named <code class="docutils literal"><span class="pre">alignments.bam</span></code>. The reads
that did not align to the reference will include <code class="docutils literal"><span class="pre">XC</span></code> attributes in the
BAM file that describe why a read did not map.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p>For more information about the RTG <code class="docutils literal"><span class="pre">map</span></code> command, refer
to <a class="reference internal" href="rtg_command_reference.html#map"><span class="std std-ref">Command Reference, map</span></a>.</p>
<p class="last">For details on RTG extensions to the BAM file format, refer to <a class="reference internal" href="appendix.html#sam-bam-file-extensions-rtg-map-command-output"><span class="std std-ref">SAM/BAM file extensions (RTG map command output)</span></a></p>
</div>
</div>
<div class="section" id="read-mapping-sensitivity-tuning">
<h3>Read mapping sensitivity tuning<a class="headerlink" href="#read-mapping-sensitivity-tuning" title="Permalink to this headline">¶</a></h3>
<p>The RTG map command uses default sensitivity settings that balance
mapping percentage and speed requirements. These settings deliver
excellent results in most cases, especially in human read sequence data
from Illumina runs with error rates of 2% or less.</p>
<p>However, some experiments demand read mapping that accommodates higher
machine error, genome mutation, or cross-species comparison. For these
situations, the investigator can set various tuning parameters to
increase the mapping percentage.</p>
<p>For reads shorter than 64 bp, RTG allows an investigator to select the
number of substitutions and indels that the <code class="docutils literal"><span class="pre">map</span></code> command will &#8220;at
least&#8221; produce. For example, using the <code class="docutils literal"><span class="pre">-a</span></code> parameter to specify the
number of allowed substitutions (i.e., mismatches) at 1, will guarantee
that the <code class="docutils literal"><span class="pre">map</span></code> command finds all alignments with at least 1
substitution.</p>
<p>For reads equal to or longer than 64 base pairs, RTG allows an
investigator to modify word and step size parameters related to the
index. These parameters are set by default to 18 or half the read
length, whichever is smaller. Decreasing the values (using <code class="docutils literal"><span class="pre">-w</span></code> for word
size and <code class="docutils literal"><span class="pre">-s</span></code> for step size) will increase the percentage of mapped
reads at the expense of additional processing time, and in the case of
step size, increased memory usage.</p>
<p>The number of mismatches threshold can be altered to increase or
decrease the number of mapped reads. Using the <code class="docutils literal"><span class="pre">--max-mated-mismatches</span></code>
parameter for example, an investigator might limit reported alignments
to only those at or lower than the given threshold value.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last">For more information about the RTG <code class="docutils literal"><span class="pre">map</span></code> command&#8217;s
sensitivity and tuning parameters, refer to <a class="reference internal" href="rtg_command_reference.html#map"><span class="std std-ref">Command Reference, map</span></a></p>
</div>
</div>
<div class="section" id="protein-search">
<h3>Protein search<a class="headerlink" href="#protein-search" title="Permalink to this headline">¶</a></h3>
<p>The <code class="docutils literal"><span class="pre">mapx</span></code> command implements a search of translated nucleotide sequence
data against one or more protein databases, with alignment sensitivity
adjusted for gaps and mismatches. With <code class="docutils literal"><span class="pre">mapx</span></code>, an investigator can sort
and classify knowns, and identify homologs and novels.</p>
<p>The <code class="docutils literal"><span class="pre">mapx</span></code> command accepts reads formatted as nucleotide data and a
reference database formatted as protein data. In a two-step process,
queries that have one or more exact matches of an <em>k</em>-mer against the
database during the matching phase are then aligned to the subject
sequence with a full edit-distance calculation using the BLOSUM62
scoring matrix.</p>
<p>The <code class="docutils literal"><span class="pre">mapx</span></code> command outputs the statistical significance of matches based
on semi-global alignments (globally across query). Reported search
results may be modified by a combination of one or more thresholds on %
identity, E value, bit score and alignment score. The output results
file is similar in construct to that reported by BLASTX.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last">For more information about the RTG <code class="docutils literal"><span class="pre">mapx</span></code> command please
refer to <a class="reference internal" href="rtg_command_reference.html#mapx"><span class="std std-ref">Command Reference, mapx</span></a></p>
</div>
</div>
<div class="section" id="protein-search-output-files">
<h3>Protein search output files<a class="headerlink" href="#protein-search-output-files" title="Permalink to this headline">¶</a></h3>
<p>The <code class="docutils literal"><span class="pre">mapx</span></code> command writes search results and a summary file in a
directory specified by the <code class="docutils literal"><span class="pre">-o</span></code> parameter at the command line. The
summary file is named <code class="docutils literal"><span class="pre">summary.txt</span></code>. There is also a file called
<code class="docutils literal"><span class="pre">progress</span></code> that can be used to monitor overall progress during a run,
and a file named <code class="docutils literal"><span class="pre">mapx.log</span></code> containing technical information that may be
useful for debugging.</p>
<p>The protein search results are written to a file named
<code class="docutils literal"><span class="pre">alignments.tsv.gz</span></code>. Each record in this results file, representing a
valid search result, is written as tab-separated fields on a single
line. The output fields are very similar to those reported by BLASTX.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last">For detailed information about the RTG <code class="docutils literal"><span class="pre">mapx</span></code> command
results file format refer to <a class="reference internal" href="appendix.html#mapx-output-file-description"><span class="std std-ref">Mapx output file description</span></a></p>
</div>
</div>
<div class="section" id="protein-search-sensitivity-tuning">
<h3>Protein search sensitivity tuning<a class="headerlink" href="#protein-search-sensitivity-tuning" title="Permalink to this headline">¶</a></h3>
<p>The RTG <code class="docutils literal"><span class="pre">mapx</span></code> command builds a set of indexes from the translated reads
and scans each query for matches according to user-specified sensitivity
settings. Sensitivity is set with two parameters. The word size (<code class="docutils literal"><span class="pre">-w</span></code> or
<code class="docutils literal"><span class="pre">--word</span></code>) parameter specifies match length. The mismatches (<code class="docutils literal"><span class="pre">-a</span></code> or
<code class="docutils literal"><span class="pre">--mismatches</span></code>) parameter specifies the number and placement of <em>k</em>-mers
across each translated query.</p>
<p>The alignment score threshold can be altered to increase or decrease the
number of mapped reads. Using the <code class="docutils literal"><span class="pre">--max-mated-score</span></code> parameter for
example, an investigator might limit reported alignments to only those
at or lower than the given threshold value.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last">For more information about the RTG <code class="docutils literal"><span class="pre">mapx</span></code> command&#8217;s
sensitivity and tuning parameters, refer to <a class="reference internal" href="appendix.html#mapx-output-file-description"><span class="std std-ref">Mapx output file description</span></a></p>
</div>
</div>
<div class="section" id="benchmarking-and-optimization-utilities">
<h3>Benchmarking and optimization utilities<a class="headerlink" href="#benchmarking-and-optimization-utilities" title="Permalink to this headline">¶</a></h3>
<p>RTG benchmarking and optimization utilities consist of simulators that
generate read and reference genome sequence data, and evaluators that
verify the accuracy of sequence search and data analysis functions.
Investigators will use these utility commands to evaluate the use of RTG
software in various read mapping and data analysis scenarios.</p>
<p>RTG provides several simulators:</p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">genomesim</span></code>
The <code class="docutils literal"><span class="pre">genomesim</span></code> command generates a reference genome with one or more
segments of varying length and a percentage mix of nucleotide values.
Use the command to create simulated genomes for benchmarking and
evaluation.</li>
<li><code class="docutils literal"><span class="pre">readsim</span></code> / <code class="docutils literal"><span class="pre">cgsim</span></code>
The <code class="docutils literal"><span class="pre">readsim</span></code> / <code class="docutils literal"><span class="pre">cgsim</span></code> commands generate synthetic read sequence
data from an input reference genome, introducing errors at a
specified rate. Use the commands to create simulated read sets for
benchmarking and evaluation.</li>
<li><code class="docutils literal"><span class="pre">popsim</span></code>, <code class="docutils literal"><span class="pre">samplesim</span></code>, <code class="docutils literal"><span class="pre">childsim</span></code>, <code class="docutils literal"><span class="pre">samplereplay</span></code>, <code class="docutils literal"><span class="pre">denovosim</span></code>
These variant simulation commands are used to create mutated genomes
from a known reference by adding variants. Use these commands to
verify accuracy of variant detection analysis software for a
particular experiment using different pipeline settings.</li>
</ul>
<p>Simulated data that is produced in SDF format can be converted into
FASTA and FASTQ format sequence files for use with other tools using the
<code class="docutils literal"><span class="pre">sdf2fasta</span></code> and <code class="docutils literal"><span class="pre">sdf2fastq</span></code> commands respectively.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last">For more information about the RTG simulation commands,
refer to <a class="reference internal" href="rtg_command_reference.html#simulation-commands"><span class="std std-ref">Simulation Commands</span></a>. Advice is
available to ensure best results. Please contact RTG technical support
for assistance.</p>
</div>
</div>
</div>
<div class="section" id="variant-detection-functions">
<h2>Variant detection functions<a class="headerlink" href="#variant-detection-functions" title="Permalink to this headline">¶</a></h2>
<p>The RTG variant detection pipeline includes commands for both sequence
and structural variation detection: <code class="docutils literal"><span class="pre">snp</span></code>, <code class="docutils literal"><span class="pre">family</span></code>, <code class="docutils literal"><span class="pre">population</span></code>,
<code class="docutils literal"><span class="pre">somatic</span></code>, <code class="docutils literal"><span class="pre">cnv</span></code> and <code class="docutils literal"><span class="pre">coverage</span></code>. The types of data available for
analysis from the RTG software pipeline include: Bayesian sequence
variant calling (<code class="docutils literal"><span class="pre">snps.vcf</span></code>), structural variation analysis
(<code class="docutils literal"><span class="pre">cnv.ratio</span></code>) and alignment coverage depth (<code class="docutils literal"><span class="pre">coverage.bed</span></code>).</p>
<div class="section" id="sequence-variation-snps-indels-and-complex-variants">
<h3>Sequence variation (SNPs, indels and complex variants)<a class="headerlink" href="#sequence-variation-snps-indels-and-complex-variants" title="Permalink to this headline">¶</a></h3>
<p>The <code class="docutils literal"><span class="pre">snp</span></code> command uses a Bayesian probability model to identify and locate
single and multiple nucleotide polymorphisms (SNPs and MNPs), indels,
and complex sequence variants. The command uses standard BAM format
files as input and reports computed posterior scores, base calls,
mapping quality, coverage depth, and supporting statistics for all
positions and for all variants. The <code class="docutils literal"><span class="pre">snp</span></code> command may be instructed to
run in either haploid or diploid calling mode, and can perform sex-aware
calling to automatically switch between haploid and diploid calling
according to sex chromosomes specified for your reference species.</p>
<p>The <code class="docutils literal"><span class="pre">snp</span></code> command calls single nucleotide polymorphisms (SNPs), multiple
nucleotide polymorphisms (MNPs), and complex regions from the sorted
chromosome-ordered gapped alignment (BAM) files. The <code class="docutils literal"><span class="pre">snp</span></code> command makes
consensus SNP and MNP calls on a diploid organism at every position
(homozygous, heterozygous, and equal) in the reference, and calls indels
and complex variants of 1-50 bp (depending on input alignments).</p>
<p>At each position in the reference, a base pair determination is made
based on statistical analysis of the accumulated read alignments,
in accordance with any priors and quality scores. The resulting
predictions and accompanying statistics are reported in industry
standard VCF format.</p>
<p>The <code class="docutils literal"><span class="pre">snps.vcf</span></code> output file reports all the called variants.
The location and type of the call, the base pairs (reference and
called), and a confidence score are present in the <code class="docutils literal"><span class="pre">snps.vcf</span></code>
output file. Additional ancillary statistics in the output describe read
alignment evidence that can be used to further evaluate confidence in the variant.
Results may be filtered (post variant calling) by posterior
scores, coverage depth, or indels, and filtered report results may be
integrated with the SNP calls themselves.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last">For more information about the SNP output data, refer to
<a class="reference internal" href="rtg_command_reference.html#map"><span class="std std-ref">Command Reference, map</span></a>, <a class="reference internal" href="rtg_command_reference.html#snp"><span class="std std-ref">Command Reference, snp</span></a> for syntax, parameters,
and usage of the <code class="docutils literal"><span class="pre">map</span></code> and <code class="docutils literal"><span class="pre">snp</span></code> commands.</p>
</div>
</div>
<div class="section" id="sequence-variation-with-mendelian-pedigree">
<h3>Sequence variation with Mendelian pedigree<a class="headerlink" href="#sequence-variation-with-mendelian-pedigree" title="Permalink to this headline">¶</a></h3>
<p>The <code class="docutils literal"><span class="pre">family</span></code> command uses Bayesian analysis and the constraints of
Mendelian inheritance to identify single and multiple nucleotide
variants in each member of a family group. It will usually yield a
better result than running the <code class="docutils literal"><span class="pre">snp</span></code> command on each individual because
the Mendelian constraints help eliminate erroneous calls.</p>
<p>Family calling is restricted to families comprising a mother, father,
and one or more sons and daughters. Family members are identified on the
command line by sample names matching those used in the input BAM files.
The family caller internally assigns the SAM records to the correct
family member based on SAM read group information. If available, it
automatically makes use of coverage and quality calibration information
computed during mapping. It automatically selects the correct
haploid/diploid calling depending on the sex of each individual.</p>
<p>The output is a multi-sample VCF file containing a call for each family
member whenever any one of the family differs from the reference genome.
Each sample reports a computed posterior, base call, and ancillary
statistics as per the <code class="docutils literal"><span class="pre">snp</span></code> command. In addition, there is an overall
posterior representing the joint likelihood of the call across all the
samples. As with the other variant detection commands, the VCF output
includes a filter column containing markers for high-coverage,
high-ambiguity, and equivalent calls. It is not guaranteed that the
resulting calls will always be Mendelian across the entire family, as
<em>de novo</em> mutations are also identified and are automatically annotated
in the output VCF.</p>
<p>The <code class="docutils literal"><span class="pre">population</span></code> command extends calling to multiple samples, which
may or may not be related according to a supplied pedigree. Mendelian
constraints are employed where appropriate, and in cases where many
unrelated samples are being called, an iterative
expectation-maximization algorithm updates Bayesian priors to give
improved accuracy compared to calling samples individually with the
<code class="docutils literal"><span class="pre">snp</span></code> command.</p>
</div>
<div class="section" id="somatic-sequence-variation">
<h3>Somatic sequence variation<a class="headerlink" href="#somatic-sequence-variation" title="Permalink to this headline">¶</a></h3>
<p>The <code class="docutils literal"><span class="pre">somatic</span></code> command uses Bayesian analysis to identify putative cancer
explanations in a tumor sample. As with the snp command, it can
identify SNPs, MNPs, indels, and complex sequence variants. It operates
on two samples, an original sample (assumed to be non-cancerous) and a
derived cancerous sample. The derived sample may be a mixture of
non-cancerous and cancerous sequence data. The samples are provided to
the somatic command in the form of BAM format files with appropriate
sample names selected via the read group mechanism.</p>
<p>The somatic caller produces a VCF file detailing putative cancer
explanations consisting of computed posterior scores, base calls, and
ancillary statistics for both input samples. The somatic caller handles
both haploid and diploid sequences and is sex aware. If available, it
automatically makes use of coverage and quality calibration information
computed during mapping.</p>
<p>By default the <code class="docutils literal"><span class="pre">snps.vcf</span></code> output file gives each variant called where
the original and derived sample differ, together with a confidence. The
file is sorted by genomic position. The same statistics reported by the
<code class="docutils literal"><span class="pre">snp</span></code> command per VCF record are listed for both samples. The filter
column contains markers for situations of high-coverage, high-ambiguity,
and equivalent calls. This column can be used to discard unwanted
results in subsequent processing.</p>
</div>
<div class="section" id="coverage-analysis">
<h3>Coverage analysis<a class="headerlink" href="#coverage-analysis" title="Permalink to this headline">¶</a></h3>
<p>The <code class="docutils literal"><span class="pre">coverage</span></code> command reports read depth across a reference genome with
smoothing options, and outputs the results in the industry standard BED
format. This can used to view histograms of mapped coverage data and gap
length distributions.</p>
<p>Use the <code class="docutils literal"><span class="pre">coverage</span></code> command as a tool to analyze mapping results and
determine how much of the genome is covered with mapping alignments, and
how many times the same location has been mapped.</p>
<p>Customizable scripts are available for enabling graphical plotting of
the coverage results using <code class="docutils literal"><span class="pre">gnuplot</span></code>.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last">For more information about the RTG coverage analysis, refer
to <a class="reference internal" href="rtg_command_reference.html#coverage"><span class="std std-ref">Command reference, coverage</span></a></p>
</div>
</div>
<div class="section" id="copy-number-variation-cnv-analysis">
<h3>Copy number variation (CNV) analysis<a class="headerlink" href="#copy-number-variation-cnv-analysis" title="Permalink to this headline">¶</a></h3>
<p>The <code class="docutils literal"><span class="pre">cnv</span></code> command identifies and reports copy number statistics that can
be used for the investigation of structural variation.</p>
<p>It is used to identify aberrational CNV region(s) or copy number
variations in a mapped read. The RTG <code class="docutils literal"><span class="pre">cnv</span></code> command identifies and
reports the copy number variation ratio between two genomes.</p>
<p>The results of CNV detection are output to a BED file format.
Customizable scripts are available for enabling graphical plotting of
the CNV results using tools such as <code class="docutils literal"><span class="pre">gnuplot</span></code>.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last">For more information about the CNV output data, refer to
<a class="reference internal" href="rtg_command_reference.html#cnv"><span class="std std-ref">Command Reference, cnv</span></a></p>
</div>
</div>
</div>
<div class="section" id="standard-input-and-output-file-formats">
<h2>Standard input and output file formats<a class="headerlink" href="#standard-input-and-output-file-formats" title="Permalink to this headline">¶</a></h2>
<p>RTG software produces alignment and data analysis results in standard
formats to allow pipeline validation and downstream analysis.</p>
<p>Table : Result file formats for validation and downstream analysis</p>
<table border="1" class="docutils">
<colgroup>
<col width="4%" />
<col width="96%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">File type</th>
<th class="head">Description and Usage</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>BAM, SAM</td>
<td>The RTG <code class="docutils literal"><span class="pre">map</span></code> and <code class="docutils literal"><span class="pre">cgmap</span></code> commands produces alignment results in the Binary Sequence Alignment/Map (BAM) format: <code class="docutils literal"><span class="pre">alignments.bam</span></code> or optionally the compressed ASCII (SAM format) equivalent <code class="docutils literal"><span class="pre">alignments.sam.gz</span></code>. This allows use of familiar pileup viewers for quick visual inspection of alignment results.</td>
</tr>
<tr class="row-odd"><td>TXT</td>
<td>Many RTG commands output summary statistics as ASCII text files.</td>
</tr>
<tr class="row-even"><td>TSV</td>
<td>Many RTG commands output results in tab separated ASCII text files. These files can typically be loaded directly into a spreadsheet viewing program like Microsoft Excel or Open Office.</td>
</tr>
<tr class="row-odd"><td>BED</td>
<td>Some RTG commands output results in standard BED formats for further analysis and reporting.</td>
</tr>
<tr class="row-even"><td>PED</td>
<td>Some RTG commands utilize standard PED format text files for supplying sample pedigree and sex information.</td>
</tr>
<tr class="row-odd"><td>VCF</td>
<td>The <code class="docutils literal"><span class="pre">snp</span></code>, <code class="docutils literal"><span class="pre">family</span></code>, <code class="docutils literal"><span class="pre">population</span></code> and <code class="docutils literal"><span class="pre">somatic</span></code> commands output results in Variant Call Format (VCF) version 4.1.</td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last">For more information about file format extensions, refer
to Appendix <a class="reference internal" href="appendix.html#rtg-output-results-file-descriptions"><span class="std std-ref">RTG output results file descriptions</span></a></p>
</div>
<div class="section" id="sam-bam-files-created-by-the-rtg-map-command">
<h3>SAM/BAM files created by the RTG map command<a class="headerlink" href="#sam-bam-files-created-by-the-rtg-map-command" title="Permalink to this headline">¶</a></h3>
<p>The Sequence Alignment/Map (SAM/BAM) format (version 1.3) is a
well-known standard for listing read alignments against reference
sequences.</p>
<p>SAM records list the mapping and alignment data for each read, ordered
by chromosome (or other DNA reference sequence) location.</p>
<p>A sample RTG SAM file is shown in the Appendix. It describes the
relationship between a read and a reference sequence, including
mismatches, insertions and deletions (indels) as determined by the RTG
<code class="docutils literal"><span class="pre">map</span></code> aligner.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">RTG mapped alignments are stored in BAM format with RTG
read IDs by default. This default can be overridden using the
<code class="docutils literal"><span class="pre">--read-names</span></code> flag or changed after processing using the RTG
<code class="docutils literal"><span class="pre">samrename</span></code> utility to label the reads with the sequence identifiers
from the original source file. For more information, refer to the SAM
1.3 nomenclature and symbols online at:
<a class="reference external" href="https://samtools.github.io/hts-specs/SAMv1.pdf">https://samtools.github.io/hts-specs/SAMv1.pdf</a></p>
</div>
<p>RTG has defined several extensions within the standard SAM/BAM format;
be sure to review the SAM/BAM format information in
<a class="reference internal" href="appendix.html#sam-bam-file-extensions-rtg-map-command-output"><span class="std std-ref">SAM/BAM file extensions (RTG map command output)</span></a> of the <em>Appendix</em> to this guide for a complete
list of all extensions and differences.</p>
<p>By default the RTG <code class="docutils literal"><span class="pre">map</span></code> command produces output as compressed binary
format BAM files but can be set to produce human readable SAM files
instead with the <code class="docutils literal"><span class="pre">--sam</span></code> flag.</p>
</div>
<div class="section" id="variant-caller-output-files">
<h3>Variant caller output files<a class="headerlink" href="#variant-caller-output-files" title="Permalink to this headline">¶</a></h3>
<p>The Variant Call Format (VCF) is a widely used standard format for
storing SNPs, MNPs and indels.</p>
<p>A sample <code class="docutils literal"><span class="pre">snps.vcf</span></code> file is provided in the Appendix as an example of
the output produced by an RTG variant calling run. Each line in a
<code class="docutils literal"><span class="pre">snps.vcf</span></code> output has tab-separated fields and represents a SNP
variation calculated from the mapped reads against the reference genome.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">RTG variant calls are stored in VCF format (version
4.1). For more information about the VCF format, refer to the
specification online at:
<a class="reference external" href="https://samtools.github.io/hts-specs/VCFv4.1.pdf">https://samtools.github.io/hts-specs/VCFv4.1.pdf</a></p>
</div>
<p>RTG employs several extensions within the standard VCF format; be sure
to review the VCF format information in <a class="reference internal" href="appendix.html#small-variant-vcf-output-file-description"><span class="std std-ref">Small-variant VCF output file description</span></a>
of the <em>Appendix</em> to this guide for a complete list of all extensions
and differences.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last">For more information about file formats, refer to the
<em>Appendix,</em> <a class="reference internal" href="appendix.html#rtg-output-results-file-descriptions"><span class="std std-ref">RTG output results file descriptions</span></a></p>
</div>
</div>
</div>
<div class="section" id="metagenomic-analysis-functions">
<h2>Metagenomic analysis functions<a class="headerlink" href="#metagenomic-analysis-functions" title="Permalink to this headline">¶</a></h2>
<p>The RTG metagenomic analysis pipeline includes commands for sample
contamination filtering, estimation of taxon abundances in a sample and
finding relationships between samples.</p>
<div class="section" id="contamination-filtering">
<h3>Contamination filtering<a class="headerlink" href="#contamination-filtering" title="Permalink to this headline">¶</a></h3>
<p>The <code class="docutils literal"><span class="pre">mapf</span></code> command is used for filtering contaminant reads from a
sample. It does this by performing alignment of the reads against a
reference of known contaminants and producing an output of the reads
that did not align successfully. A common use for this is to remove
human DNA from a bacterial sample taken from a body site.</p>
</div>
<div class="section" id="taxon-abundance-breakdown">
<h3>Taxon abundance breakdown<a class="headerlink" href="#taxon-abundance-breakdown" title="Permalink to this headline">¶</a></h3>
<p>The <code class="docutils literal"><span class="pre">species</span></code> command is used to find the abundances of taxa within a
given sample. This is accomplished by analyzing reference genome
alignment data made with a metagenomic reference database of known
organisms. It produces output in which each taxon is given a fraction
representing its abundance in the sample with upper and lower bounds and
a value indicating the confidence that the taxon is actually present in
the sample. An HTML report allows interactive examination of the
abundances at different taxonomic levels.</p>
</div>
<div class="section" id="sample-relationships">
<h3>Sample relationships<a class="headerlink" href="#sample-relationships" title="Permalink to this headline">¶</a></h3>
<p>The <code class="docutils literal"><span class="pre">similarity</span></code> command is used to find relationships between sample
read sets. It does this by examining <em>k</em>-mer word frequencies and the
intersections between sets of reads. This results in the output of a
similarity matrix, a principal component analysis and nearest neighbor
trees in the Newick and phyloXML formats.</p>
</div>
<div class="section" id="functional-protein-analysis">
<h3>Functional protein analysis<a class="headerlink" href="#functional-protein-analysis" title="Permalink to this headline">¶</a></h3>
<p>The <code class="docutils literal"><span class="pre">mapx</span></code> command is used to perform a translated nucleotide search of
short reads against a reference protein database. This results in an
output similar to that reported by BLASTX.</p>
</div>
</div>
<div class="section" id="pipelines">
<h2>Pipelines<a class="headerlink" href="#pipelines" title="Permalink to this headline">¶</a></h2>
<p>Included in the RTG release are some pipeline commands which perform
simple end-to-end tasks using other RTG commands. These pipelines use
mostly default settings for each of the commands called, and are meant
as a guideline to building more complex end-to-end pipelines using our
tools. The metagenomic pipeline commands are:</p>
<ul class="simple">
<li>species composition (<code class="docutils literal"><span class="pre">composition-meta-pipeline</span></code>)</li>
<li>functional protein analysis (<code class="docutils literal"><span class="pre">functional-meta-pipeline</span></code>)</li>
<li>species composition and functional protein analysis
(<code class="docutils literal"><span class="pre">composition-functional-meta-pipeline</span></code>).</li>
</ul>
<p>For detailed information about individual pipeline commands see
<a class="reference internal" href="rtg_command_reference.html#pipeline-commands"><span class="std std-ref">Pipeline Commands</span></a></p>
</div>
<div class="section" id="parallel-processing">
<h2>Parallel processing<a class="headerlink" href="#parallel-processing" title="Permalink to this headline">¶</a></h2>
<p>The comparison of genomic variation on a large scale in real time
demands parallel processing capability. Parallel processing of gapped
alignments and variant detection is recommended by RTG because it
significantly reduces wall clock time.</p>
<p>RTG software includes key features that make it easier for a person to
prepare a job for parallel processing. First, RTG mapping commands can
be performed on a subset of a large file or set of files either by using
the <code class="docutils literal"><span class="pre">--start-read</span></code> and <code class="docutils literal"><span class="pre">--end-read</span></code> parameters or for commands that do
not support this, by using <code class="docutils literal"><span class="pre">sdfsplit</span></code> to break a large SDF into smaller
pieces. Second, the data analysis commands accept multiple alignment
files as input from the command. Third, many RTG commands take a
<code class="docutils literal"><span class="pre">--region</span></code> or <code class="docutils literal"><span class="pre">--bed-regions</span></code> parameter to allow breaking up tasks into
pieces across the reference genome.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last">See <a class="reference internal" href="rtg_command_reference.html#rtg-command-reference"><span class="std std-ref">RTG Command Reference</span></a>
for command-specific details, <a class="reference internal" href="administration.html#administration-capacity-planning"><span class="std std-ref">Administration &amp; Capacity Planning</span></a>
for detailed information about estimating the number of
multi-core servers needed (capacity planning), and <a class="reference internal" href="appendix.html#parallel-processing-approach"><span class="std std-ref">Parallel processing approach</span></a> for a
deeper discussion of compute cluster operations.</p>
</div>
</div>
<div class="section" id="installation-and-deployment">
<h2>Installation and deployment<a class="headerlink" href="#installation-and-deployment" title="Permalink to this headline">¶</a></h2>
<p>RTG is a self-contained tool that sets minimal expectations on the
environment in which it is placed. It comes with the application
components it needs to execute completely, yet performance can be
enhanced with some simple modifications to the deployment configuration.
This section provides guidelines for installing and creating an optimal
configuration, starting from a typical recommended system.</p>
<p>RTG software pipeline runs in a wide range of computing environments
from dual-core processor laptops to compute clusters with racks of dual
processor quad core server nodes. However, internal human genome
analysis benchmarks suggest the use of six server nodes of the
configuration shown in below.</p>
<p>Table : Recommended system requirements</p>
<table border="1" class="docutils">
<colgroup>
<col width="23%" />
<col width="77%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>Processor</td>
<td>Intel Core i7-2600</td>
</tr>
<tr class="row-even"><td>Memory</td>
<td>48 GB RAM DDR3</td>
</tr>
<tr class="row-odd"><td>Disk</td>
<td>5 TB, 7200 RPM (prefer SAS disk)</td>
</tr>
</tbody>
</table>
<p>RTG Software can be run as a Java JAR file, but platform specific
wrapper scripts are supplied to provide improved pipeline ergonomics.
Instructions for a quick start installation are provided here.</p>
<p>For further information about setting up per-machine configuration
files, please see the <code class="docutils literal"><span class="pre">README.txt</span></code> contained in the distribution zip
file (a copy is also included in this manual&#8217;s appendix).</p>
<div class="section" id="quick-start-instructions">
<h3>Quick start instructions<a class="headerlink" href="#quick-start-instructions" title="Permalink to this headline">¶</a></h3>
<p>These instructions are intended for an individual to install and operate
the RTG software without the need to establish root / administrator
privileges.</p>
<p>RTG software is delivered in a compressed zip file, such as:
<code class="docutils literal"><span class="pre">rtg-core-3.3.zip</span></code>. Unzip this file to begin installation.</p>
<p>Linux and Windows distributions include a Java Virtual Machine (JVM)
version 1.8 that has undergone quality assurance testing. RTG may be
used on other operating systems for which a JVM version 1.8 or higher is
available, such as MacOS X or Solaris, by using the &#8216;no-jre&#8217;
distribution.</p>
<p>RTG for Java is delivered as a Java application accessed via executable
wrapper script (<code class="docutils literal"><span class="pre">rtg</span></code> on UNIX systems, <code class="docutils literal"><span class="pre">rtg.bat</span></code> on Windows) that allows
a user to customize initial memory allocation and other configuration
options. It is recommended that these wrapper scripts be used rather
than directly executing the Java JAR.</p>
<p>Here are platform-specific instructions for RTG deployment.</p>
<p>Linux/MacOS X:</p>
<ul>
<li><p class="first">Unzip the RTG distribution to the desired location.</p>
</li>
<li><p class="first">If your installation requires a license file (<code class="docutils literal"><span class="pre">rtg-license.txt</span></code>),
copy the license file provided by Real Time Genomics into the RTG
distribution directory.</p>
</li>
<li><dl class="first docutils">
<dt>In a terminal, cd to the installation directory and test for success</dt>
<dd><p class="first last">by entering <code class="docutils literal"><span class="pre">./rtg</span> <span class="pre">version</span></code></p>
</dd>
</dl>
</li>
<li><p class="first">On MacOS X, depending on your operating system version and
configuration regarding unsigned applications, you may encounter the
error message:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>-bash: rtg: /usr/bin/env: bad interpreter: Operation not permitted
</pre></div>
</div>
<p>If this occurs, you must clear the OS X quarantine attribute with the
command:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ xattr -d com.apple.quarantine rtg
</pre></div>
</div>
</li>
<li><p class="first">The first time rtg is executed you will be prompted with some
questions to customize your installation. Follow the prompts.</p>
</li>
<li><p class="first">Enter <code class="docutils literal"><span class="pre">./rtg</span> <span class="pre">help</span></code> for a list of rtg commands. Help for any
individual command is available using the <code class="docutils literal"><span class="pre">--help</span></code> flag, e.g.:
<code class="docutils literal"><span class="pre">./rtg</span> <span class="pre">format</span> <span class="pre">--help</span></code></p>
</li>
<li><p class="first">By default, RTG software scripts establish a memory space of 90% of
the available RAM - this is automatically calculated. One may
override this limit in the <code class="docutils literal"><span class="pre">rtg.cfg</span></code> settings file or on a per-run
basis by supplying <code class="docutils literal"><span class="pre">RTG_MEM</span></code> as an environment variable or as the
first program argument, e.g.: <code class="docutils literal"><span class="pre">./rtg</span> <span class="pre">RTG_MEM=48g</span> <span class="pre">map</span></code></p>
</li>
<li><p class="first">[OPTIONAL] If you will be running RTG on multiple machines and would
like to customize settings on a per-machine basis, copy <code class="docutils literal"><span class="pre">rtg.cfg</span></code> to
<code class="docutils literal"><span class="pre">/etc/rtg.cfg</span></code>, editing per-machine settings appropriately (requires
root privileges).  An alternative that does not require root
privileges is to copy <code class="docutils literal"><span class="pre">rtg.cfg</span></code> to <code class="docutils literal"><span class="pre">rtg.HOSTNAME.cfg</span></code>, editing
per-machine settings appropriately, where <code class="docutils literal"><span class="pre">HOSTNAME</span></code> is the short host
name output by the command <code class="docutils literal"><span class="pre">hostname</span> <span class="pre">-s</span></code></p>
</li>
</ul>
<p>Windows:</p>
<ul class="simple">
<li>Unzip the RTG distribution to the desired location.</li>
<li>If your installation requires a license, copy the license file
provided by Real Time Genomics (<code class="docutils literal"><span class="pre">rtg-license.txt</span></code>) into the RTG
distribution directory.</li>
<li>Test for success by entering <code class="docutils literal"><span class="pre">rtg</span> <span class="pre">version</span></code> at the command line. The
first time RTG is executed you will be prompted with some questions
to customize your installation. Follow the prompts.</li>
<li>Enter <code class="docutils literal"><span class="pre">rtg</span> <span class="pre">help</span></code> for a list of rtg commands. Help for any
individual command is available using the <code class="docutils literal"><span class="pre">--help</span></code> flag, e.g.:
<code class="docutils literal"><span class="pre">./rtg</span> <span class="pre">format</span> <span class="pre">--help</span></code></li>
<li>By default, RTG software scripts establish a memory space of 90% of
the available RAM - this is automatically calculated. One may
override this limit by setting the <code class="docutils literal"><span class="pre">RTG_MEM</span></code> variable in the
<code class="docutils literal"><span class="pre">rtg.bat</span></code> script or as an environment variable.</li>
</ul>
</div>
<div class="section" id="license-management">
<h3>License Management<a class="headerlink" href="#license-management" title="Permalink to this headline">¶</a></h3>
<p>Commercial distributions of RTG products require the presence of a valid
license key file for operation.</p>
<p>The license key file must be located in the same directory as the RTG
executable. The license enables the execution of a particular command
set for the purchased product(s) and features.</p>
<p>A license key allows flexible use of the RTG package on any node or CPU
core.</p>
<p>To view the current license features at the command prompt, enter:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg license
</pre></div>
</div>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last">For more data center deployment and instructions for
editing scripts, see <a class="reference internal" href="administration.html#administration-capacity-planning"><span class="std std-ref">Administration &amp; Capacity Planning</span></a>.</p>
</div>
</div>
</div>
<div class="section" id="technical-assistance-and-support">
<h2>Technical assistance and support<a class="headerlink" href="#technical-assistance-and-support" title="Permalink to this headline">¶</a></h2>
<p>For assistance with any technical or conceptual issue that may arise
during use of the RTG product, contact Real Time Genomics Technical
Support via email at <code class="docutils literal"><span class="pre">support&#64;realtimegenomics.com</span></code></p>
<p>In addition, a discussion group is available at:
<a class="reference external" href="https://groups.google.com/a/realtimegenomics.com/forum/#!forum/rtg-users">https://groups.google.com/a/realtimegenomics.com/forum/#!forum/rtg-users</a></p>
<p>A low-traffic announcements-only group is available at:
<a class="reference external" href="https://groups.google.com/a/realtimegenomics.com/forum/#!forum/rtg-announce">https://groups.google.com/a/realtimegenomics.com/forum/#!forum/rtg-announce</a></p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="rtg_command_reference.html" title="RTG Command Reference"
             >next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Table Of Contents"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">RTG Core Operations Manual v3.10</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2017, Real Time Genomics.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.5.3.
    </div>
  </body>
</html>