

<!doctype html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>RTG product usage - baseline progressions &#8212; RTG Core Operations Manual v3.7</title>
    
    <link rel="stylesheet" href="_static/bizstyle.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '3.7',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/bizstyle.js"></script>
    <link rel="top" title="RTG Core Operations Manual v3.7" href="index.html" />
    <link rel="next" title="Administration &amp; Capacity Planning" href="administration.html" />
    <link rel="prev" title="RTG Command Reference" href="rtg_command_reference.html" />
    <meta name="viewport" content="width=device-width,initial-scale=1.0">
    <!--[if lt IE 9]>
    <script type="text/javascript" src="_static/css3-mediaqueries.js"></script>
    <![endif]-->
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="administration.html" title="Administration &amp; Capacity Planning"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="rtg_command_reference.html" title="RTG Command Reference"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">RTG Core Operations Manual v3.7</a> &#187;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="index.html">
              <img class="logo" src="_static/html_logo.png" alt="Logo"/>
            </a></p>
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">RTG product usage - baseline progressions</a><ul>
<li><a class="reference internal" href="#rtg-mapping-and-sequence-variant-detection">RTG mapping and sequence variant detection</a><ul>
<li><a class="reference internal" href="#task-1-format-reference-data">Task 1 - Format reference data</a></li>
<li><a class="reference internal" href="#task-2-format-read-data">Task 2 - Format read data</a></li>
<li><a class="reference internal" href="#task-3-map-reads-to-the-reference-genome">Task 3 - Map reads to the reference genome</a></li>
<li><a class="reference internal" href="#task-4-view-evaluate-mapping-performance">Task 4 - View &amp; evaluate mapping performance</a></li>
<li><a class="reference internal" href="#task-5-generate-and-review-coverage-information">Task 5 - Generate and review coverage information</a></li>
<li><a class="reference internal" href="#task-6-call-sequence-variants">Task 6 - Call sequence variants</a></li>
<li><a class="reference internal" href="#task-7-report-copy-number-variation-statistics">Task 7 - Report copy number variation statistics</a></li>
</ul>
</li>
<li><a class="reference internal" href="#modifications-to-support-complete-genomics-inc-reads">Modifications to support Complete Genomics, Inc reads</a><ul>
<li><a class="reference internal" href="#task-2-format-read-data-complete-genomics">Task 2 - Format read data (Complete Genomics)</a></li>
<li><a class="reference internal" href="#task-3-map-reads-to-the-reference-genome-complete-genomics">Task 3 - Map reads to the reference genome (Complete Genomics)</a></li>
</ul>
</li>
<li><a class="reference internal" href="#modifications-to-support-processing-exome-sequencing">Modifications to support processing exome sequencing</a></li>
<li><a class="reference internal" href="#modifications-to-support-variant-calling-with-multiple-platforms">Modifications to support variant calling with multiple platforms</a></li>
<li><a class="reference internal" href="#modifications-to-support-sex-aware-mapping-and-variant-calling">Modifications to support sex-aware mapping and variant calling</a><ul>
<li><a class="reference internal" href="#task-1-format-reference-data-with-reference-txt">Task 1 - Format reference data with reference.txt</a></li>
<li><a class="reference internal" href="#task-3-map-reads-to-the-reference-genome-with-sex-information">Task 3 - Map reads to the reference genome with sex information</a></li>
<li><a class="reference internal" href="#task-6-call-sequence-variants-with-sex-information">Task 6 - Call sequence variants with sex information</a></li>
</ul>
</li>
<li><a class="reference internal" href="#modifications-to-support-joint-family-variant-calling">Modifications to support joint family variant calling</a><ul>
<li><a class="reference internal" href="#task-3-map-reads-to-the-reference-genome-family-calling">Task 3 - Map reads to the reference genome (family calling)</a></li>
<li><a class="reference internal" href="#task-6-call-sequence-variants-with-family-settings">Task 6 - Call sequence variants with family settings</a></li>
</ul>
</li>
<li><a class="reference internal" href="#modifications-to-support-joint-population-variant-calling">Modifications to support joint population variant calling</a><ul>
<li><a class="reference internal" href="#task-3-map-reads-to-the-reference-genome-population-calling">Task 3 – Map reads to the reference genome (population calling)</a></li>
<li><a class="reference internal" href="#task-6-call-population-sequence-variants">Task 6 - Call population sequence variants</a></li>
</ul>
</li>
<li><a class="reference internal" href="#create-and-use-population-priors-in-variant-calling">Create and use population priors in variant calling</a><ul>
<li><a class="reference internal" href="#task-1-produce-population-priors-file">Task 1 - Produce population priors file</a></li>
<li><a class="reference internal" href="#task-2-run-variant-calling-using-population-priors">Task 2 - Run variant calling using population priors</a></li>
</ul>
</li>
<li><a class="reference internal" href="#somatic-variant-detection-in-cancer">Somatic variant detection in cancer</a><ul>
<li><a class="reference internal" href="#task-1-format-reference-data-somatic">Task 1 - Format reference data (Somatic)</a></li>
<li><a class="reference internal" href="#task-2-format-read-data-somatic">Task 2 - Format read data (Somatic)</a></li>
<li><a class="reference internal" href="#task-3-map-tumor-and-normal-sample-reads-against-the-reference-genome">Task 3 - Map tumor and normal sample reads against the reference genome</a></li>
<li><a class="reference internal" href="#task-4-call-somatic-variants">Task 4 - Call somatic variants</a></li>
<li><a class="reference internal" href="#using-site-specific-somatic-priors">Using site-specific somatic priors</a></li>
</ul>
</li>
<li><a class="reference internal" href="#avr-scoring-using-hapmap-for-model-building">AVR scoring using HAPMAP for model building</a><ul>
<li><a class="reference internal" href="#task-1-create-training-data">Task 1 - Create training data</a></li>
<li><a class="reference internal" href="#task-2-build-and-check-avr-model">Task 2 - Build and check AVR model</a></li>
<li><a class="reference internal" href="#task-3-use-avr-model">Task 3 - Use AVR model</a></li>
<li><a class="reference internal" href="#task-4-install-avr-model">Task 4 - Install AVR model</a></li>
</ul>
</li>
<li><a class="reference internal" href="#rtg-structural-variant-detection">RTG structural variant detection</a><ul>
<li><a class="reference internal" href="#task-1-format-reference-and-read-data">Task 1 - Format reference and read data</a></li>
<li><a class="reference internal" href="#task-2-map-reads-against-a-reference-genome">Task 2 - Map reads against a reference genome</a></li>
<li><a class="reference internal" href="#task-4-find-structural-variants-with-sv">Task 4 - Find structural variants with sv</a></li>
<li><a class="reference internal" href="#task-5-find-structural-variants-with-discord">Task 5 - Find structural variants with discord</a></li>
</ul>
</li>
<li><a class="reference internal" href="#ion-torrent-bacterial-mapping-and-sequence-variant-detection">Ion Torrent bacterial mapping and sequence variant detection</a><ul>
<li><a class="reference internal" href="#task-1-format-bacterial-reference-data-ion-torrent">Task 1 - Format bacterial reference data (Ion Torrent)</a></li>
<li><a class="reference internal" href="#task-2-format-read-data-ion-torrent">Task 2 - Format read data (Ion Torrent)</a></li>
<li><a class="reference internal" href="#task-3-map-ion-torrent-reads-to-the-reference-genome">Task 3 - Map Ion Torrent reads to the reference genome</a></li>
<li><a class="reference internal" href="#task-4-call-sequence-variants-in-haploid-mode">Task 4 - Call sequence variants in haploid mode</a></li>
</ul>
</li>
<li><a class="reference internal" href="#rtg-contaminant-filtering">RTG contaminant filtering</a><ul>
<li><a class="reference internal" href="#task-1-format-reference-data-contaminant-filtering">Task 1 - Format reference data (contaminant filtering)</a></li>
<li><a class="reference internal" href="#task-2-format-read-data-contaminant-filtering">Task 2 - Format read data (contaminant filtering)</a></li>
<li><a class="reference internal" href="#task-3-run-contamination-filter">Task 3 - Run contamination filter</a></li>
<li><a class="reference internal" href="#task-4-manage-filtered-reads">Task 4 - Manage filtered reads</a></li>
</ul>
</li>
<li><a class="reference internal" href="#rtg-translated-protein-searching">RTG translated protein searching</a><ul>
<li><a class="reference internal" href="#task-1-format-protein-data-set">Task 1 - Format protein data set</a></li>
<li><a class="reference internal" href="#task-2-format-dna-read-set">Task 2 - Format DNA read set</a></li>
<li><a class="reference internal" href="#task-3-search-against-protein-data-set">Task 3 - Search against protein data set</a></li>
</ul>
</li>
<li><a class="reference internal" href="#rtg-species-frequency-estimation">RTG species frequency estimation</a><ul>
<li><a class="reference internal" href="#task-1-format-reference-data-species">Task 1 - Format reference data (species)</a></li>
<li><a class="reference internal" href="#task-2-format-read-data-species">Task 2 - Format read data (species)</a></li>
<li><a class="reference internal" href="#task-3-run-contamination-filter-optional">Task 3 - Run contamination filter (optional)</a></li>
<li><a class="reference internal" href="#task-4-map-metagenomic-reads-against-bacterial-database">Task 4 - Map metagenomic reads against bacterial database</a></li>
<li><a class="reference internal" href="#task-5-run-species-estimator">Task 5 - Run species estimator</a></li>
</ul>
</li>
<li><a class="reference internal" href="#rtg-sample-similarity">RTG sample similarity</a><ul>
<li><a class="reference internal" href="#task-1-prepare-read-sets">Task 1 - Prepare read sets</a></li>
<li><a class="reference internal" href="#task-2-generate-read-set-name-map">Task 2 - Generate read set name map</a></li>
<li><a class="reference internal" href="#task-3-run-similarity-tool">Task 3 - Run similarity tool</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="rtg_command_reference.html"
                        title="previous chapter">RTG Command Reference</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="administration.html"
                        title="next chapter">Administration &amp; Capacity Planning</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/product_usage_wrapper.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="rtg-product-usage-baseline-progressions">
<span id="baseline-product-usage"></span><h1>RTG product usage - baseline progressions<a class="headerlink" href="#rtg-product-usage-baseline-progressions" title="Permalink to this headline">¶</a></h1>
<p>This chapter provides baseline task progressions that describe expected
use of the product for data analysis.</p>
<div class="section" id="rtg-mapping-and-sequence-variant-detection">
<h2>RTG mapping and sequence variant detection<a class="headerlink" href="#rtg-mapping-and-sequence-variant-detection" title="Permalink to this headline">¶</a></h2>
<p>Use the following set of ordered tasks to detect all sequence variants
between a reference genome and a sequenced DNA sample. This set of tasks
steps through the main functionality of the RTG variant detection
software pipeline: generating and evaluating gapped alignments, testing
coverage depth, calling sequence variants (SNPs, MNPs and indels), and
analyzing copy number variation between samples.</p>
<p>The following example supports the steps typical to single sample human
whole genome analysis in which high throughput sequencing with Illumina
sequencing systems has generated reads at 25 to 30× coverage. This
example does not demonstrate exome variant calling which is described in
<a class="reference internal" href="#modifications-to-support-processing-exome-sequencing"><span class="std std-ref">Modifications to support processing exome sequencing</span></a>, the
automatic sex-aware calling capabilities which are described in
<a class="reference internal" href="#modifications-to-support-sex-aware-mapping-and-variant-calling"><span class="std std-ref">Modifications to support sex-aware mapping and variant calling</span></a>,
or joint variant calling capabilities which are described in
<a class="reference internal" href="#modifications-to-support-joint-family-variant-calling"><span class="std std-ref">Modifications to support joint family variant calling</span></a> and
<a class="reference internal" href="#modifications-to-support-joint-population-variant-calling"><span class="std std-ref">Modifications to support joint population variant calling</span></a></p>
<p>Data</p>
<p>The baseline uses actual data downloaded from National Center for
Biotechnology Information (NCBI) databases. The reference sequence is
human build 37 (hg19, Feb 2009) downloaded from
<a class="reference external" href="ftp://hgdownload.cse.ucsc.edu/goldenPath/hg19/chromosomes/">ftp://hgdownload.cse.ucsc.edu/goldenPath/hg19/chromosomes/</a>.</p>
<p>The paired end read data comes from a human genome sequencing of an
African female (Yoruba) individual (HapMap: NA19240). This data can be
found in the NCBI Short Read Archive (SRA).</p>
<p>Table : Overview of basic pipeline tasks</p>
<table border="1" class="docutils">
<colgroup>
<col width="5%" />
<col width="20%" />
<col width="12%" />
<col width="64%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="2">Task</th>
<th class="head">Command &amp; Utilities</th>
<th class="head">Purpose</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>1</td>
<td>Format reference data</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">format</span></code></td>
<td>Convert reference sequence from FASTA files to RTG Sequence Data Format (SDF)</td>
</tr>
<tr class="row-odd"><td>2</td>
<td>Format read data</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">format</span></code></td>
<td>Convert read sequence from FASTA or FASTQ files to RTG Sequence Data Format (SDF)</td>
</tr>
<tr class="row-even"><td>3</td>
<td>Map reads against a reference genome</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">map</span></code></td>
<td>Generate read alignments against a given reference, and report in a BAM file for downstream analysis</td>
</tr>
<tr class="row-odd"><td>4</td>
<td>View alignment results</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">samstats</span></code></td>
<td>Evaluate alignments and determine if the mapping job(s) should be repeated with different settings</td>
</tr>
<tr class="row-even"><td>5</td>
<td>Generate coverage information</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">coverage</span></code></td>
<td>Run the coverage command to generate coverage breadth and depth statistics</td>
</tr>
<tr class="row-odd"><td>6</td>
<td>Call sequence variants</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">snp</span></code></td>
<td>Detect SNPs, MNPs, and indels in a sample relative to a reference genome</td>
</tr>
<tr class="row-even"><td>7</td>
<td>Report copy number variation</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">cnv</span></code></td>
<td>Count copy number variation (CNV) ratios between two genomes, and present results that support structural variation analysis</td>
</tr>
</tbody>
</table>
<div class="section" id="task-1-format-reference-data">
<h3>Task 1 - Format reference data<a class="headerlink" href="#task-1-format-reference-data" title="Permalink to this headline">¶</a></h3>
<p>RTG tools require a conversion of reference genome from FASTA files into
the RTG SDF format. This task will be completed with the <code class="docutils literal"><span class="pre">format</span></code>
command.  The conversion will create an SDF directory containing the
reference genome.</p>
<p>For variant calling and CNV analysis, the reference genome employed must
have chromosome-based coordinates (as opposed to contig-based
coordinates), such as hg18 or GRCh37/hg19.</p>
<p>First, observe a typical genome reference with multiple chromosomes
stored in compressed FASTA format.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ ls -l /data/human/hg19/
43026389 Mar 21 2009 chr10.fa.gz
42966674 Mar 21 2009 chr11.fa.gz
42648875 Mar 21 2009 chr12.fa.gz
31517348 Mar 21 2009 chr13.fa.gz
28970334 Mar 21 2009 chr14.fa.gz
26828094 Mar 21 2009 chr15.fa.gz
25667827 Mar 21 2009 chr16.fa.gz
25139792 Mar 21 2009 chr17.fa.gz
24574571 Mar 21 2009 chr18.fa.gz
17606811 Mar 21 2009 chr19.fa.gz
73773666 Mar 21 2009 chr1.fa.gz
19513342 Mar 21 2009 chr20.fa.gz
11549785 Mar 21 2009 chr21.fa.gz
11327826 Mar 21 2009 chr22.fa.gz
78240395 Mar 21 2009 chr2.fa.gz
64033758 Mar 21 2009 chr3.fa.gz
61700369 Mar 21 2009 chr4.fa.gz
58378199 Mar 21 2009 chr5.fa.gz
54997756 Mar 21 2009 chr6.fa.gz
50667196 Mar 21 2009 chr7.fa.gz
46889258 Mar 21 2009 chr8.fa.gz
39464200 Mar 21 2009 chr9.fa.gz
    5537 Mar 21 2009 chrM.fa.gz
49278128 Mar 21 2009 chrX.fa.gz
 8276338 Mar 21 2009 chrY.fa.gz
</pre></div>
</div>
<p>Now, use the <code class="docutils literal"><span class="pre">format</span></code> command to convert multiple input files into a
single SDF directory for the reference genome.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg format -o hg19 /data/human/hg19/chrM.fa.gz \
  /data/human/hg19/chr[1-9].fa.gz \
  /data/human/hg19/chr1[0-9].fa.gz \
  /data/human/hg19/chr2[0-9].fa.gz \
  /data/human/hg19/chrX.fa.gz \
  /data/human/hg19/chrY.fa.gz
</pre></div>
</div>
<p>This takes the human reference FASTA files and creates a directory
called hg19 containing the SDF, with chromosomes ordered in the standard
UCSC ordering.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">When formatting a reference genome, the <code class="docutils literal"><span class="pre">format</span></code> command
will automatically recognize several common human reference genomes
and install a <code class="docutils literal"><span class="pre">reference.txt</span></code> configuration file.  For reference
genomes which are not recognized, if you plan on performing sex-aware
mapping and variant calling, you should copy or create an appropriate
<code class="docutils literal"><span class="pre">reference.txt</span></code> file into the SDF directory. See <a class="reference internal" href="#modifications-to-support-sex-aware-mapping-and-variant-calling"><span class="std std-ref">Modifications to support sex-aware mapping and variant calling</span></a> and <a class="reference internal" href="appendix.html#rtg-reference-file-format"><span class="std std-ref">RTG reference file format</span></a></p>
</div>
<p>You can use the <code class="docutils literal"><span class="pre">sdfstats</span></code> command to show statistics for
your reference SDF, including the individual sequence lengths.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg sdfstats --lengths hg19
Type               : DNA
Number of sequences: 25
Maximum length     : 249250621
Minimum length     : 16571
Sequence names     : yes
Sex metadata       : yes
Taxonomy metadata  : no
N                  : 234350281
A                  : 844868045
C                  : 585017944
G                  : 585360436
T                  : 846097277
Total residues     : 3095693983
Residue qualities  : no

Sequence lengths:
chrM    16571
chr1    249250621
chr2    243199373
chr3    198022430
chr4    191154276
chr5    180915260
chr6    171115067
chr7    159138663
chr8    146364022
chr9    141213431
chr10   135534747
chr11   135006516
chr12   133851895
chr13   115169878
chr14   107349540
chr15   102531392
chr16   90354753
chr17   81195210
chr18   78077248
chr19   59128983
chr20   63025520
chr21   48129895
chr22   51304566
chrX    155270560
chrY    59373566
</pre></div>
</div>
<p>Note the presence of reference chromosome metadata on the line <code class="docutils literal"><span class="pre">Sex</span> <span class="pre">metadata</span></code>.</p>
</div>
<div class="section" id="task-2-format-read-data">
<h3>Task 2 - Format read data<a class="headerlink" href="#task-2-format-read-data" title="Permalink to this headline">¶</a></h3>
<p>RTG mapping tools accept input read sequence data either in the RTG SDF
format or other common sequence formats such as FASTA, FASTQ, or
SAM/BAM. There are pros and cons as to whether to perform an initial
format of the read sequence data to RTG SDF:</p>
<ul class="simple">
<li>Pre-formatting requires an extra one-off workflow step (the <code class="docutils literal"><span class="pre">format</span></code>
command), whereas native input file formats are directly accepted by
many RTG commands.</li>
<li>Pre-formatting requires extra disk space for the SDF (although these
can be deleted after processing if required).</li>
<li>With pre-formatting, decompression, parsing and error checking raw
files is carried out only once, whereas native formats require this
processing each time.</li>
<li>Pre-formatting permits random access to individual sequences or
blocks of sequences, whereas with native formats, the whole file
leading up to the region of interest must also be decompressed, and
parsed.</li>
<li>Pre-formatting permits loading of sequence data, sequence names, and
sequence quality values independently, allowing reduced RAM use
during mapping</li>
</ul>
<p>Thus, pre-formatting read sequence data can result in lower overall
resource requirements (and faster throughput) than processing native
file formats directly.</p>
<p>In this example we will be converting read sequence data from FASTA or
FASTQ files into the RTG SDF format. This task will be completed with
the <code class="docutils literal"><span class="pre">format</span></code> commands. The conversion will create one or more SDF
directories for the reads.</p>
<p>Take a paired set of reads in FASTQ format and convert it into RTG data
format (SDF). This example shows one run of data, taking as input both
left and right mate pairs from the same run.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ mkdir sample_NA19240
$ rtg format -f fastq -q sanger -o sample_NA19240/SRR002978 \
  -l /data/reads/NA19240/SRR002978_1.fq \
  -r /data/reads/NA19240/SRR002978_2.fq \
  --sam-rg &quot;@RG\\tID:SRR002978\\tSM:NA19240\\tPL:ILLUMINA&quot;
</pre></div>
</div>
<p>This creates a directory named SRR002978 with two subdirectories, named
left and right. Use the <code class="docutils literal"><span class="pre">sdfstats</span></code> command to verify this step.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg sdfstats sample_NA19240/SRR002978
</pre></div>
</div>
<p>It is good practice to ensure the output BAM files contain tracking
information regarding the read set. This can be achieved by storing the
tracking information in the reads SDF by using the <code class="docutils literal"><span class="pre">--sam-rg</span></code> flag to
provide a SAM-formatted read group header line. This information will
automatically be used during mapping to enable automatic creation of
calibration files that are used to perform base-quality recalibration
during variant calling. In addition, the sample field (<code class="docutils literal"><span class="pre">SM</span></code>) is used to
segregate samples when using multi-sample variant calling commands such
as <code class="docutils literal"><span class="pre">family</span></code>, <code class="docutils literal"><span class="pre">population</span></code>, and <code class="docutils literal"><span class="pre">somatic</span></code>. The header should specify at
least the <code class="docutils literal"><span class="pre">ID</span></code>, <code class="docutils literal"><span class="pre">SM</span></code> and <code class="docutils literal"><span class="pre">PL</span></code> fields for the read group. For more
details see <a class="reference internal" href="appendix.html#using-sam-bam-read-groups-in-rtg-map"><span class="std std-ref">Using SAM/BAM Read Groups in RTG map</span></a>.</p>
<p>If the tracking information is not specified when formatting the reads,
it can be set during mapping instead using the <code class="docutils literal"><span class="pre">--sam-rg</span></code> flag.</p>
<p>Repeat for all available read data associated with the sample to be
processed. This example shows how this can be done with the format
command in a loop.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ for left_fq in /data/reads/NA19240/*_1.fq; do
$ right_fq=${left_fq/_1.fq/_2.fq}
$ lane_id=$(basename ${left_fq/_1.fq})
$ rtg format -f fastq -q sanger -o ${lane_id} -l ${left_fq} -r ${right_fq} \
  --sam-rg &quot;@RG\tID:${lane_id}\tSM:NA19240\tPL:ILLUMINA&quot;
$ done
</pre></div>
</div>
</div>
<div class="section" id="task-3-map-reads-to-the-reference-genome">
<h3>Task 3 - Map reads to the reference genome<a class="headerlink" href="#task-3-map-reads-to-the-reference-genome" title="Permalink to this headline">¶</a></h3>
<p>Map the sequence reads against the human reference genome to generate
alignments in the BAM format (Binary Sequence Alignment/Map file
format).</p>
<p>The RTG <code class="docutils literal"><span class="pre">map</span></code> command provides multiple tuning parameters to adjust
sensitivity and selectivity at the read mapping stage. In general, whole
genome mapping strategies aim to capture the highest number of reads
possessing variant data and which map with a high degree of specificity
to the human genome.</p>
<p>Depending on the downstream analysis required, the mapping may be
adjusted for restricting alignments to unique genomic positions or for
allowing ambiguous mappings at equivalent regions throughout the
genome. This example will show the baseline for human genome analysis.</p>
<p>By default, the <code class="docutils literal"><span class="pre">map</span></code> command will report all mated pairs and unmated
reads that map to the reference genome.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg map -i sample_NA19240/SRR002978 -t hg19 -o \
  map_sample_NA19240/SRR002978
</pre></div>
</div>
<p>Multiple files are written to the output directory of the mapping
run. By default the <code class="docutils literal"><span class="pre">alignments.bam</span></code> file is produced and a BAM index
is automatically created to permit efficient extraction of mappings
within particular genomic regions. This behavior is necessary for
subsequent analysis of the mappings, but can be performed manually using
the index command.  The additional file <code class="docutils literal"><span class="pre">alignments.bam.calibration</span></code>
is also created, containing metadata regarding coverage and error rates
which help to provide more accurate variant calling.</p>
<p>The size of the job should be tuned to the available memory of the
server node. You can perform mapping in segments by using the
<code class="docutils literal"><span class="pre">--start-read</span></code> and <code class="docutils literal"><span class="pre">--end-read</span></code> flags during mapping.  Currently, a
48GB RAM system as specified in the technical requirements section can
process 100M reads in about an hour. The following example would work to
map a data set containing just over 100M reads in batches of 10M.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ for ((j=0;j&lt;10;j++)); do
$ rtg map -i sample1-100M -t hg19 -o map_sample1-$j
  --start-read=$[j*10000000] --end-read=$[(j+1)*10000000]
$ done
</pre></div>
</div>
<p>Note that each of these runs can be executed independently of the
others. This allows parallel processing in a compute cluster that can
reduce wall clock time.</p>
<p>In a parallel compute cluster special consideration is needed with
respect to where data resides during the mapping process. Reading and
writing data from and to a single networked disk partition may result in
undesirable I/O performance characteristics depending on the size and
structure of the compute cluster. One way to minimize the adverse
affects of I/O limitations is to separate input data sets and map output
directories, storing them on different network disk partitions.</p>
<p>Using the <code class="docutils literal"><span class="pre">--tempdir</span></code> flag allows the <code class="docutils literal"><span class="pre">map</span></code> command to use a
directory other than the output directory to store temporary files that
are output during the mapping process. The size of the temporary files
is the same as the total size of files in the map output directory after
processing has finished. The following example shows how to modify the
above example to place outputs on a separate partition to the inputs.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ mkdir /partition2/map_sample_NA19240
$ for ((j=0;j&lt;9;j++)); do
$ rtg map -i /partition1/map_sample_NA19240/SRR002978
  -t /partition1/hg19
  -o /partition2/map_sample_NA19240/SRR002978-$j
  --start-read=$[j*1000000] --end-read=$[(j+1)*1000000]
  --tempdir /partition3/map_temp_SRR002978-$j
$ done
</pre></div>
</div>
</div>
<div class="section" id="task-4-view-evaluate-mapping-performance">
<h3>Task 4 - View &amp; evaluate mapping performance<a class="headerlink" href="#task-4-view-evaluate-mapping-performance" title="Permalink to this headline">¶</a></h3>
<p>An <code class="docutils literal"><span class="pre">alignments.bam</span></code> file can be viewed directly using the RTG
<code class="docutils literal"><span class="pre">extract</span></code> command in conjunction with a shell command such as <code class="docutils literal"><span class="pre">less</span></code>
to quickly inspect the results.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg extract map_sample_NA19240/SRR002978-0/alignments.bam | less -S
</pre></div>
</div>
<p>Since the mappings are indexed by default, it is also possible to view
mappings corresponding to particular genomic regions. For example:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg extract map_sample_NA19240/SRR002978-0/alignments.bam &#39;chr6:1000000-5000000&#39;\
  | less -S
</pre></div>
</div>
<p>The <code class="docutils literal"><span class="pre">map</span></code> command also creates a simple HTML summary report containing
information about mapping counts, alignment score distribution,
paired-end insert size distribution, etc. This can be viewed in your web
browser. For example:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ firefox map_sample_NA19240/SRR002978-0/index.html
</pre></div>
</div>
<p>For more detailed summary statistics, use the <code class="docutils literal"><span class="pre">samstats</span></code> command. This
example will report information such as total records, number unmapped,
specific details about the mate pair reads, and distributions for
alignment scores, insert sizes and read hits.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg samstats -t hg19 map_sample_NA19240/SRR002978-0/alignments.bam \
  -r /partition1/map_sample_NA19240/SRR002978 --distributions
</pre></div>
</div>
</div>
<div class="section" id="task-5-generate-and-review-coverage-information">
<h3>Task 5 - Generate and review coverage information<a class="headerlink" href="#task-5-generate-and-review-coverage-information" title="Permalink to this headline">¶</a></h3>
<p>For human genomic analysis, it’s important to have sufficient coverage
over the entire genome to detect variants accurately and minimize false
negative calls. The <code class="docutils literal"><span class="pre">coverage</span></code> command provides detailed statistics for
depth of coverage and gap size distribution. If coverage proves to be
inadequate or spotty, one may elect to remap the data with different
sensitivity tuning or rerun the sample with different sequencing
technology.</p>
<p>This example shows the <code class="docutils literal"><span class="pre">coverage</span></code> command used with all alignments, both
mated and unmated, for the entire sample. The <code class="docutils literal"><span class="pre">-s</span></code> flag is used to
introduce smoothing of the data, by default the data will not be
smoothed. While you can supply the <code class="docutils literal"><span class="pre">coverage</span></code> command with the names of
BAM files individually on the command line, this becomes unwieldy when
the mapping has been carried out in many smaller jobs. In this command
we will use the <code class="docutils literal"><span class="pre">-I</span></code> flag to supply a text file containing the names of
all the mapping output BAM files, one per line. One example way to
create this file is with the following command (assuming all your
mapping runs have used a common root directory):</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ find /partition2/map_sample_NA19240 -name &quot;*.bam&quot; &gt; sample_NA19240-bam-files
$ rtg coverage -s 20 -o cov_sample_NA19240 -t hg19 -I sample_NA19240-bam-files
</pre></div>
</div>
<p>By default the <code class="docutils literal"><span class="pre">coverage</span></code> command will generate a BED formatted file
containing regions of similar coverage. This BED file can be loaded into
a genome browser to visualize the coverage, and may also be examined on
the command line. For example:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg extract cov_sample_NA19240/coverage.bed.gz &#39;chr6:1000000-5000000&#39; | less
</pre></div>
</div>
<p>The <code class="docutils literal"><span class="pre">coverage</span></code> command also creates a simple HTML summary report
containing graphs of the depth of coverage distribution and cumulative
depth of coverage. This can be viewed in your web browser. For example:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ firefox cov_sample_NA19240/index.html
</pre></div>
</div>
</div>
<div class="section" id="task-6-call-sequence-variants">
<h3>Task 6 - Call sequence variants<a class="headerlink" href="#task-6-call-sequence-variants" title="Permalink to this headline">¶</a></h3>
<p>The <code class="docutils literal"><span class="pre">snp</span></code> command detects sequence variants, given adequate but not
excessively high coverage of reads against the reference genome. As with
the <code class="docutils literal"><span class="pre">coverage</span></code> command, we can supply a file containing a list of all
the needed input files. In this case the mapping calibration files will
be automatically detected at the location of the BAM files themselves
and will be used in order to enable base quality recalibration during
variant calling.</p>
<p>This example takes all available BAM files for the sample and calls SNP,
MNP and indel sequence variants.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg snp -o snp_sample_NA19240 -t hg19 -I sample_NA19240-bam-files
</pre></div>
</div>
<p>This command will perform the Bayesian variant calling and will use a
default AVR model for scoring variant call quality. If you have a more
appropriate model available, you should supply this with the
<code class="docutils literal"><span class="pre">--avr-model</span></code> flag. RTG supplies some pre-built models which work well
in a wide variety of cases, and includes tools for building custom AVR
models. See <a class="reference internal" href="#avr-scoring-using-hapmap-for-model-building"><span class="std std-ref">AVR scoring using HAPMAP for model building</span></a> for more information on
building custom models.</p>
<p>The <code class="docutils literal"><span class="pre">snp</span></code> command will output variant call summary information upon
completion, and this is also available in the output directory in the
file <code class="docutils literal"><span class="pre">summary.txt</span></code>.</p>
<p>One can quickly view and process the resulting compressed output file
with the <code class="docutils literal"><span class="pre">zless</span></code> shell command.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ zless -S snp_sample_NA19240/snps.vcf.gz
</pre></div>
</div>
<p>Since the variant calls are compressed and indexed by default, it is
also possible to view calls corresponding to particular genomic regions.
For example:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg extract snp_sample_NA19240/snps.vcf.gz &#39;chr6:1000000-5000000&#39; | less -S
</pre></div>
</div>
<p>The following example would work to call variants using a separate job
for each reference sequence separately. Each of these runs could be
executed on independent nodes of a cluster in order to reduce wall clock
time.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ for seq in M {1..22} X Y; do
$ rtg snp -o snp_sample_chr&quot;$seq&quot; -t hg19 -I sample_NA19240-bam-files \
  --region chr&quot;$seq&quot;
$ done
</pre></div>
</div>
<p>After the separate variant calling jobs complete, the VCF files for each
chromosome can be combined into a single file if desired by using a
<code class="docutils literal"><span class="pre">vcfmerge</span></code> command such as:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg vcfmerge -o snp_sample_NA19240.vcf.gz snp_sample_chr*/snps.vcf.gz
</pre></div>
</div>
<p>The individual <code class="docutils literal"><span class="pre">snp</span></code> commands will have output summary information about
the variant calls made in each job. Combined summary information can be
output for the merged VCF file with the <code class="docutils literal"><span class="pre">vcfstats</span></code> command.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg vcfstats snp_sample_NA19240.vcf.gz
</pre></div>
</div>
<p>Simple filtering of variants can be applied using the <code class="docutils literal"><span class="pre">vcffilter</span></code>
command. For example, filtering calls by genotype quality or AVR score
can be accomplished with a command such as:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg vcffilter --min-genotype-quality 50 -i snp_sample_NA19240.vcf.gz \
  -o filtered_sample_NA19240.vcf.gz
</pre></div>
</div>
<p>In this case, any variants failing the filter will be removed.
Alternatively, failing variants can be kept but marked with a custom VCF
<code class="docutils literal"><span class="pre">FILTER</span></code> field, such as:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg vcffilter --fail LOW-GQ --min-genotype-quality 50 \
  -i snp_sample_NA19240.vcf.gz -o filtered_sample_NA19240.vcf.gz
</pre></div>
</div>
</div>
<div class="section" id="task-7-report-copy-number-variation-statistics">
<h3>Task 7 - Report copy number variation statistics<a class="headerlink" href="#task-7-report-copy-number-variation-statistics" title="Permalink to this headline">¶</a></h3>
<p>With two genome samples, one can compare the relative depth of coverage
by region to identify copy number variation ratios that may indicate
structural variation.</p>
<p>For a second genome called <code class="docutils literal"><span class="pre">sample2</span></code>, repeat tasks 2 through 6 with a
related read data set. For example, you might have two samples from a
cancer patient, one from a germline cell and another from a tumor.</p>
<p>Run the <code class="docutils literal"><span class="pre">cnv</span></code> command with the default bucket size of 100, this is the
number of nucleotides for which to average the coverage.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg cnv -o cnv_s1_s2 -I sample1-map-files -J sample2-map-files
</pre></div>
</div>
<p>View the resulting output as a set of records that show cnv ration at
locations across the genome, where the locations are defined by the
bucket size.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ zless cnv_s1_/cnv.txt.gz
</pre></div>
</div>
<p>For deeper investigation, contact Real Time Genomics technical support
for extensible reporting scripts specific to copy number variation
reporting.</p>
</div>
</div>
<div class="section" id="modifications-to-support-complete-genomics-inc-reads">
<h2>Modifications to support Complete Genomics, Inc reads<a class="headerlink" href="#modifications-to-support-complete-genomics-inc-reads" title="Permalink to this headline">¶</a></h2>
<p>RTG supports the custom read structure of Complete Genomics Inc data.
Use the same set of ordered tasks as in with the addition of the
following differences to detect all sequence variants between a
reference genome and a sequenced DNA sample.</p>
<p>Table : Overview of Complete Genomics pipeline tasks</p>
<table border="1" class="docutils">
<colgroup>
<col width="3%" />
<col width="23%" />
<col width="13%" />
<col width="60%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="2">Task</th>
<th class="head">Command &amp; Utilities</th>
<th class="head">Purpose</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>2</td>
<td>Format read data</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">cg2sdf</span></code></td>
<td>Convert read sequence from CGI files to RTG Sequence Data Format (SDF)</td>
</tr>
<tr class="row-odd"><td>3</td>
<td>Map reads against a reference genome</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">cgmap</span></code></td>
<td>Generate read alignments against a given reference, and report in a BAM file for downstream analysis</td>
</tr>
</tbody>
</table>
<div class="section" id="task-2-format-read-data-complete-genomics">
<h3>Task 2 - Format read data (Complete Genomics)<a class="headerlink" href="#task-2-format-read-data-complete-genomics" title="Permalink to this headline">¶</a></h3>
<p>Variant detection requires a conversion of read sequence data from CGI
files into the RTG SDF format. This task will be completed <code class="docutils literal"><span class="pre">cg2sdf</span></code>. The
conversion will create one or more SDF directories for the reads.</p>
<p>Take a CGI format reads file and convert it into RTG sequence data
format (SDF). This example shows one run of data, taking as input a CGI
read data file. It is important when mapping Complete Genomics reads,
that your read group information should set the platform field (<code class="docutils literal"><span class="pre">PL</span></code>)
appropriately. For version 1 reads (35 base-pair reads), the platform
should be <code class="docutils literal"><span class="pre">COMPLETE</span></code>. For version 2 reads (29 or 30 base-pair reads)
the platform should be <code class="docutils literal"><span class="pre">COMPLETEGENOMICS</span></code>.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ mkdir CGsample_NA19240
$ rtg cg2sdf -o CGsample_NA19240/GS002290 /data/reads/cg/NA19240/GS002290.tsv \
  --sam-rg &quot;@RG\tID:GS002290\tSM:NA19240\tPL:COMPLETE&quot;
</pre></div>
</div>
<p>This creates a directory named GS00290 with two subdirectories, named
left and right. Use the <code class="docutils literal"><span class="pre">sdfstats</span></code> command to verify this step.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg sdfstats CGsample_NA19240/GS002290
</pre></div>
</div>
</div>
<div class="section" id="task-3-map-reads-to-the-reference-genome-complete-genomics">
<span id="cgmap-task3"></span><h3>Task 3 - Map reads to the reference genome (Complete Genomics)<a class="headerlink" href="#task-3-map-reads-to-the-reference-genome-complete-genomics" title="Permalink to this headline">¶</a></h3>
<p>Map the sequence reads against the reference genome to generate
alignments in the BAM format (Binary Sequence Alignment/Map file
format).</p>
<p>The RTG <code class="docutils literal"><span class="pre">cgmap</span></code> command provides the means to map Complete Genomics
reads to a reference genome. By default the <code class="docutils literal"><span class="pre">cgmap</span></code> command will report
all mated pairs and unmated reads that map to the reference genome.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ mkdir cgmap_sample_NA19240
$ rtg cgmap -i CGsample_NA19240/GS002290 -t hg19 \
  -o cgmap_sample_NA19240/GS002290 –-mask cg1
</pre></div>
</div>
<p>As with the <code class="docutils literal"><span class="pre">map</span></code> command, you can use the <code class="docutils literal"><span class="pre">--start-read</span></code> and
<code class="docutils literal"><span class="pre">--end-read</span></code> flags to perform the mapping in smaller sections if
required.</p>
<p>The selection of the <code class="docutils literal"><span class="pre">--mask</span></code> parameter depends on the length of the
Complete Genomics reads to be mapped. The <code class="docutils literal"><span class="pre">cg1</span></code> and <code class="docutils literal"><span class="pre">cg1-fast</span></code> masks are
appropriate for 35 base-pair reads and the <code class="docutils literal"><span class="pre">cg2</span></code> mask is appropriate for
29 or 30 base-pair reads. See <a class="reference internal" href="rtg_command_reference.html#cgmap"><span class="std std-ref">cgmap</span></a> for more
detail.</p>
</div>
</div>
<div class="section" id="modifications-to-support-processing-exome-sequencing">
<h2>Modifications to support processing exome sequencing<a class="headerlink" href="#modifications-to-support-processing-exome-sequencing" title="Permalink to this headline">¶</a></h2>
<p>During mapping RTG automatically creates calibration files containing
information about base qualities, average coverage levels etc. This
calibration information is utilized during variant calling to give more
accurate results and to determine when coverage levels are abnormally
high. When processing exome data, it is important that this calibration
information should only be computed for mappings within the exome
capture regions, otherwise the computed coverage levels will be much
lower than actual. This can result in RTG discarding many variant calls
as being over-coverage. The recommended workflow for exome processing is
to supply a BED file describing the exome regions at the same time as
mapping, to ensure appropriate calibration is computed.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg map --bed-regions exome-regions.bed -i sample_NA19240/SRR002978 -t hg19 \
  -o map_sample_NA19240/SRR002978 -a 2 -b 1
</pre></div>
</div>
<p>It is also recommended to supply the BED file describing the exome
regions during variant calling to filter the output produced to be
within the exome regions. There are two approaches here. The first is to
instruct the variant caller to only perform variant calling at sites
within the exome regions by supplying the <code class="docutils literal"><span class="pre">--bed-regions</span></code> option:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg snp –bed-regions exome-regions.bed -o snp_sample_NA19240 -t hg19 \
  -I sample_NA19240-bam-files
</pre></div>
</div>
<p>This approach is computationally more efficient. However, if calls at
off-target sites are potentially of interest, a second approach is to
call variants at all sites but automatically filter variants as being
off-target in the VCF <code class="docutils literal"><span class="pre">FILTER</span></code> field by using the <code class="docutils literal"><span class="pre">--filter-bed</span></code> option:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg snp --filter-bed exome-regions.bed -o snp_sample_NA19240 -t hg19 \
  -I sample_NA19240-bam-files
</pre></div>
</div>
<p>Note that the exome capture BED file must correspond to the correct
reference you are mapping and calling against. You may need to run the
BED file supplied by your sequencing vendor through a lift-over tool if
the reference genome versions differ.</p>
<p>If you do not have a BED file that specifies the exome capture regions
for your reference genome, you should supply an additional flag during
variant calling to indicate an appropriate coverage threshold for
over-coverage situations. In this case, the variant command would look
like this:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg snp --max-coverage N -o snp_sample_NA19240 -t hg19 \
  -I sample_NA19240-bam-files
</pre></div>
</div>
<p>Where <code class="docutils literal"><span class="pre">N</span></code> is typically 5 times the expected coverage level within exome
regions.</p>
</div>
<div class="section" id="modifications-to-support-variant-calling-with-multiple-platforms">
<h2>Modifications to support variant calling with multiple platforms<a class="headerlink" href="#modifications-to-support-variant-calling-with-multiple-platforms" title="Permalink to this headline">¶</a></h2>
<p>When calling variants, the <code class="docutils literal"><span class="pre">snp</span></code> command will automatically set
machine error calibration parameters according to the platform (<code class="docutils literal"><span class="pre">PL</span></code>
attribute) specified in the SAM read group. The values currently
recognized by the <code class="docutils literal"><span class="pre">snp</span></code> command for this attribute are <code class="docutils literal"><span class="pre">ILLUMINA</span></code>,
<code class="docutils literal"><span class="pre">LS454</span></code>, <code class="docutils literal"><span class="pre">IONTORRENT</span></code>, <code class="docutils literal"><span class="pre">COMPLETE</span></code>, and <code class="docutils literal"><span class="pre">COMPLETEGENOMICS</span></code>. As
long as each mapping run has set a read group with an appropriate
platform attribute, data from multiple platforms can be supplied to the
<code class="docutils literal"><span class="pre">snp</span></code> command. For example, to call variants using output from the
mapping runs in the previous two sections:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ cat sample_NA19240-map-files sample_NA19240_cgmap-files \
  &gt; sample_NA19240-all-map-files
$ rtg snp -o snp_multiplatform -t hg19 -I sample_NA19240-all-map-files
</pre></div>
</div>
</div>
<div class="section" id="modifications-to-support-sex-aware-mapping-and-variant-calling">
<h2>Modifications to support sex-aware mapping and variant calling<a class="headerlink" href="#modifications-to-support-sex-aware-mapping-and-variant-calling" title="Permalink to this headline">¶</a></h2>
<p>RTG features the ability to adjust mapping and variant calling according
to the sex of the individual that has been sequenced. During mapping,
reads will only be mapped against chromosomes that are present in that
individuals genome (for example, for a female individual, reads will not
be mapped to the Y chromosome). Similarly, sex-aware variant calling
will automatically determine when to switch between diploid and haploid
prediction models.</p>
<p>This section builds on the tasks listed in to perform sex-aware mapping
and variant calling for a single individual. RTG is also capable of
performing simultaneous joint calling of multiple family members with
the <code class="docutils literal"><span class="pre">family</span></code> command, described in more detail in
<a class="reference internal" href="#modifications-to-support-joint-family-variant-calling"><span class="std std-ref">Modifications to support joint family variant calling</span></a></p>
<p>Table : Overview of sex-aware pipeline tasks</p>
<table border="1" class="docutils">
<colgroup>
<col width="7%" />
<col width="27%" />
<col width="17%" />
<col width="50%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="2">Task</th>
<th class="head">Command &amp; Utilities</th>
<th class="head">Purpose</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>1</td>
<td>Format reference data</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">format</span></code></td>
<td>Create reference specification file</td>
</tr>
<tr class="row-odd"><td>3</td>
<td>Map reads to reference genome</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">map</span></code></td>
<td>Perform sex-aware mapping to the reference</td>
</tr>
<tr class="row-even"><td>6</td>
<td>Call sequence variants</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">snp</span></code></td>
<td>Perform sex-aware variant calls relative to the reference</td>
</tr>
</tbody>
</table>
<div class="section" id="task-1-format-reference-data-with-reference-txt">
<span id="sex-aware-task1"></span><h3>Task 1 - Format reference data with reference.txt<a class="headerlink" href="#task-1-format-reference-data-with-reference-txt" title="Permalink to this headline">¶</a></h3>
<p>Sex-aware processing requires some extra information to be placed in the
reference SDF that specifies which chromosomes are present in each sex,
and the expected ploidy of each chromosome. When formatting the
reference genome as described in <a class="reference internal" href="#task-1-format-reference-data"><span class="std std-ref">Task 1 - Format reference data</span></a>,
the reference genome may have been automatically identified. If not,
create a <code class="docutils literal"><span class="pre">reference.txt</span></code> in the reference SDF directory by copying an
appropriate one of the examples provided in the <code class="docutils literal"><span class="pre">scripts</span></code> subdirectory
of the RTG distribution (in this case corresponding to hg19):</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ cp /path/to/rtg/scripts/hg19.example.reference.txt hg19/reference.txt
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">If your reference genome is for a species other than
human or if it uses different chromosome naming conventions, the
<code class="docutils literal"><span class="pre">reference.txt</span></code> file you create will need to be adjusted. See
<a class="reference internal" href="appendix.html#rtg-reference-file-format"><span class="std std-ref">RTG reference file format</span></a> for more information
about the <code class="docutils literal"><span class="pre">reference.txt</span></code> file format.</p>
</div>
<p>After placing this file in the reference SDF directory, you can use the
<code class="docutils literal"><span class="pre">sdfstats</span></code> command to verify the chromosome listing for each sex.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg sdfstats hg19 --sex male --sex female --sex either
Type               : DNA
Number of sequences: 25
Maximum length     : 249250621
Minimum length     : 16571
Sequence names     : yes
Sex metadata       : yes
Taxonomy metadata  : no
N : 234350281
A : 844868045
C : 585017944
G : 585360436
T : 846097277
Total residues     : 3095693983
Residue qualities  : no
Sequences for sex=MALE:
chrM POLYPLOID circular 16571
chr1 DIPLOID linear 249250621
chr2 DIPLOID linear 243199373
chr3 DIPLOID linear 198022430
chr4 DIPLOID linear 191154276
chr5 DIPLOID linear 180915260
chr6 DIPLOID linear 171115067
chr7 DIPLOID linear 159138663
chr8 DIPLOID linear 146364022
chr9 DIPLOID linear 141213431
chr10 DIPLOID linear 135534747
chr11 DIPLOID linear 135006516
chr12 DIPLOID linear 133851895
chr13 DIPLOID linear 115169878
chr14 DIPLOID linear 107349540
chr15 DIPLOID linear 102531392
chr16 DIPLOID linear 90354753
chr17 DIPLOID linear 81195210
chr18 DIPLOID linear 78077248
chr19 DIPLOID linear 59128983
chr20 DIPLOID linear 63025520
chr21 DIPLOID linear 48129895
chr22 DIPLOID linear 51304566
chrX HAPLOID linear 155270560 ~=chrY
chrX:60001-2699520 chrY:10001-2649520
chrX:154931044-155260560 chrY:59034050-59363566
chrY HAPLOID linear 59373566 ~=chrX
chrX:60001-2699520 chrY:10001-2649520
chrX:154931044-155260560 chrY:59034050-59363566
Sequences for sex=FEMALE:
chrM POLYPLOID circular 16571
chr1 DIPLOID linear 249250621
chr2 DIPLOID linear 243199373
chr3 DIPLOID linear 198022430
chr4 DIPLOID linear 191154276
chr5 DIPLOID linear 180915260
chr6 DIPLOID linear 171115067
chr7 DIPLOID linear 159138663
chr8 DIPLOID linear 146364022
chr9 DIPLOID linear 141213431
chr10 DIPLOID linear 135534747
chr11 DIPLOID linear 135006516
chr12 DIPLOID linear 133851895
chr13 DIPLOID linear 115169878
chr14 DIPLOID linear 107349540
chr15 DIPLOID linear 102531392
chr16 DIPLOID linear 90354753
chr17 DIPLOID linear 81195210
chr18 DIPLOID linear 78077248
chr19 DIPLOID linear 59128983
chr20 DIPLOID linear 63025520
chr21 DIPLOID linear 48129895
chr22 DIPLOID linear 51304566
chrX DIPLOID linear 155270560
chrY NONE linear 59373566
Sequences for sex=EITHER:
chrM POLYPLOID circular 16571
chr1 DIPLOID linear 249250621
chr2 DIPLOID linear 243199373
chr3 DIPLOID linear 198022430
chr4 DIPLOID linear 191154276
chr5 DIPLOID linear 180915260
chr6 DIPLOID linear 171115067
chr7 DIPLOID linear 159138663
chr8 DIPLOID linear 146364022
chr9 DIPLOID linear 141213431
chr10 DIPLOID linear 135534747
chr11 DIPLOID linear 135006516
chr12 DIPLOID linear 133851895
chr13 DIPLOID linear 115169878
chr14 DIPLOID linear 107349540
chr15 DIPLOID linear 102531392
chr16 DIPLOID linear 90354753
chr17 DIPLOID linear 81195210
chr18 DIPLOID linear 78077248
chr19 DIPLOID linear 59128983
chr20 DIPLOID linear 63025520
chr21 DIPLOID linear 48129895
chr22 DIPLOID linear 51304566
chrX DIPLOID linear 155270560
chrY DIPLOID linear 59373566
</pre></div>
</div>
<p>If you receive an error message, check the names of the chromosomes in
the <code class="docutils literal"><span class="pre">reference.txt</span></code> file match those used in the reference SDF.</p>
</div>
<div class="section" id="task-3-map-reads-to-the-reference-genome-with-sex-information">
<span id="sex-aware-task3"></span><h3>Task 3 - Map reads to the reference genome with sex information<a class="headerlink" href="#task-3-map-reads-to-the-reference-genome-with-sex-information" title="Permalink to this headline">¶</a></h3>
<p>The only alteration to mapping is that the sex of the individual should
be specified using the <code class="docutils literal"><span class="pre">--sex</span></code> flag (or the <code class="docutils literal"><span class="pre">--pedigree</span></code> flag). For
example, continuing with our earlier NA19240 examples:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ mkdir map_sample_NA19240
$ rtg map -i sample_NA19240/SRR002978 -t hg19 -o map_sample_NA19240/SRR002978 \
  --sex female
</pre></div>
</div>
<p>or for Complete Genomics reads:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ mkdir cgmap_sample_NA19240
$ rtg cgmap -i CGsample_NA19240/GS002290 -t hg19 -o cgmap_sample_NA19240/GS002290 \
  --sex female –-mask cg1
</pre></div>
</div>
<p>These commands will result in no mappings being made to the Y
chromosome. If you were mapping the father of the YRI trio, NA19239, you
would specify <code class="docutils literal"><span class="pre">--sex=male</span></code> instead. When no <code class="docutils literal"><span class="pre">--sex</span></code> flag is supplied,
this is equivalent to specifying <code class="docutils literal"><span class="pre">--sex=either</span></code>.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">When performing many mapping runs for several samples,
more streamlined processing can be achieved by using the <code class="docutils literal"><span class="pre">--pedigree</span></code>
option to specify the sample sex via a pedigree PED file. See
<a class="reference internal" href="#task-3-map-reads-to-the-reference-genome-family-calling"><span class="std std-ref">Task 3 - Map reads to the reference genome (family calling)</span></a> for more information.</p>
</div>
</div>
<div class="section" id="task-6-call-sequence-variants-with-sex-information">
<span id="sex-aware-task6"></span><h3>Task 6 - Call sequence variants with sex information<a class="headerlink" href="#task-6-call-sequence-variants-with-sex-information" title="Permalink to this headline">¶</a></h3>
<p>Performing sex-aware variant calling is just as simple as sex-aware
mapping, simply specify the appropriate <code class="docutils literal"><span class="pre">--sex</span></code> flag when running the
<code class="docutils literal"><span class="pre">snp</span></code> command.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg snp -o snp_sample_NA19240 -t hg19 --sex female -I sample_NA19240-map-files
</pre></div>
</div>
<p>Inspecting the output VCF for this run will show that no variants have
been called for the Y chromosome. If you have carried out sex-aware
mapping for the father of the trio, NA19239, the following command would
perform sex-aware variant calling.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg snp -o snp_sample_NA19239 -t hg19 --sex male -I sample_NA19239-map-files
</pre></div>
</div>
<p>Inspection of the output VCF for this run will show haploid variant
calls for the X and Y chromosomes.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg extract snp_sample_NA19239/snps.vcf.gz chrX | head
$ rtg extract snp_sample_NA19239/snps.vcf.gz chrY | head
</pre></div>
</div>
<p>RTG also supports automatic handling of pseudoautosomal regions and will
produce haploid or diploid calls within the X PAR regions as appropriate
for the sex of the individual.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">When performing many <code class="docutils literal"><span class="pre">snp</span></code> runs for several samples, more
streamlined processing can be achieved by using the <code class="docutils literal"><span class="pre">--pedigree</span></code> option
to specify the sample sex via a pedigree PED file. See
<a class="reference internal" href="#task-3-map-reads-to-the-reference-genome-family-calling"><span class="std std-ref">Task 3 - Map reads to the reference genome (family calling)</span></a> for more information.</p>
</div>
</div>
</div>
<div class="section" id="modifications-to-support-joint-family-variant-calling">
<h2>Modifications to support joint family variant calling<a class="headerlink" href="#modifications-to-support-joint-family-variant-calling" title="Permalink to this headline">¶</a></h2>
<p>This section builds on the tasks listed in and to extend variant calling
to joint calling of multiple members of a Mendelian family.</p>
<p>Table : Overview of family variant calling pipeline tasks</p>
<table border="1" class="docutils">
<colgroup>
<col width="6%" />
<col width="22%" />
<col width="14%" />
<col width="59%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="2">Task</th>
<th class="head">Command &amp; Utilities</th>
<th class="head">Purpose</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>3</td>
<td>Map reads to reference genome</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">map</span></code></td>
<td>Perform sex-aware mapping to the reference</td>
</tr>
<tr class="row-odd"><td>6</td>
<td>Call sequence variants</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">family</span></code></td>
<td>Perform sex-aware joint variant calls relative to the reference on a Mendelian family</td>
</tr>
</tbody>
</table>
<div class="section" id="task-3-map-reads-to-the-reference-genome-family-calling">
<span id="join-family-task3"></span><h3>Task 3 - Map reads to the reference genome (family calling)<a class="headerlink" href="#task-3-map-reads-to-the-reference-genome-family-calling" title="Permalink to this headline">¶</a></h3>
<p>For optimal results the mapping stage should be performed as described
in <a class="reference internal" href="#task-3-map-reads-to-the-reference-genome-with-sex-information"><span class="std std-ref">Task 3 - Map reads to the reference genome with sex information</span></a>. Also ensure that when mapping each individual they are
given a unique sample ID and that all reads for that individual have the
same sample ID (The sample ID is specified as the <code class="docutils literal"><span class="pre">SM</span></code> part of the
<code class="docutils literal"><span class="pre">--sam-rg</span></code> flag either during the formatting of the reads or for the
<code class="docutils literal"><span class="pre">map</span></code> command explicitly).</p>
<p>For example when mapping the trio NA19238, NA19239 and NA19240, we shall
give the mappings for NA19238 <code class="docutils literal"><span class="pre">--sex=female</span></code> option and the sample ID
NA19238:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ mkdir map_sample_NA19238
$ rtg map -i sample_NA19238/SRR002998 -t hg19 -o map_sample_NA19238/SRR002998 \
  --sex female --sam-rg &quot;@RG\tID:SRR002998\tSM:NA19238\tPL:ILLUMINA&quot;
</pre></div>
</div>
<p>For NA19239 <code class="docutils literal"><span class="pre">--sex=male</span></code> and sample id NA19239:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ mkdir map_sample_NA19239
$ rtg map -i sample_NA19239/SRR002879 -t hg19 -o map_sample_NA19239/SRR002879 \
  --sex male --sam-rg &quot;@RG\tID:SRR002879\tSM:NA19239\tPL:ILLUMINA&quot;
</pre></div>
</div>
<p>And lastly for NA19240 <code class="docutils literal"><span class="pre">--sex=female</span></code> option and the sample ID NA19240:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ mkdir map_sample_NA19240
$ rtg map -i sample_NA19240/SRR002978 -t hg19 -o map_sample_NA19240/SRR002978 \
  --sex female --sam-rg &quot;@RG\tID:SRR002978\tSM:NA19240\tPL:ILLUMINA&quot;
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Per-sample sex information can also be
specified using a pedigree PED file. See <a class="reference internal" href="#population-map"><span class="std std-ref">Task 3 - Map reads to the reference genome</span></a></p>
</div>
</div>
<div class="section" id="task-6-call-sequence-variants-with-family-settings">
<span id="join-family-task6"></span><h3>Task 6 - Call sequence variants with family settings<a class="headerlink" href="#task-6-call-sequence-variants-with-family-settings" title="Permalink to this headline">¶</a></h3>
<p>The <code class="docutils literal"><span class="pre">family</span></code> command is invoked similarly to the <code class="docutils literal"><span class="pre">snp</span></code> command except
you need to specify the way the samples relate to each other. This is
done by specifying the appropriate sample ID given during mapping to the
command line flags <code class="docutils literal"><span class="pre">--mother</span></code>, <code class="docutils literal"><span class="pre">--father</span></code>, <code class="docutils literal"><span class="pre">--daughter</span></code> and <code class="docutils literal"><span class="pre">--son</span></code>.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The RTG <code class="docutils literal"><span class="pre">family</span></code> command only supports a basic
family relationship of a mother, father and one or more children, either
daughters or sons. For other pedigrees, use the <code class="docutils literal"><span class="pre">population</span></code> command.</p>
</div>
<p>To run the <code class="docutils literal"><span class="pre">family</span></code> command on the trio of NA19238 (mother), NA19239
(father) and NA19240 (daughter) you need to provide all the mapping
files for the samples. The mapping calibration files will be
automatically detected at the locations of the mapping files. To specify
these in a file list for input you could run:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ find /partition2/map_trio -name &quot;alignments.bam&quot; &gt; map_trio-bam-files
</pre></div>
</div>
<p>To run the <code class="docutils literal"><span class="pre">family</span></code> command you then specify the sample ID for each member
of the trio to the appropriate flag.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg family --mother NA19238 --father NA19239 --daughter NA19240 -t hg19 \
  -o trio_variants -I map_trio-bam-files
</pre></div>
</div>
<p>Examining the <code class="docutils literal"><span class="pre">snps.vcf.gz</span></code> file in the output directory will show
individual columns for the variants of each family member. For more
details about the VCF output file see <a class="reference internal" href="appendix.html#vcf-output-file-description"><span class="std std-ref">VCF output file description</span></a></p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Per-family relationship information can also
be specified using a pedigree PED file. See
<a class="reference internal" href="#population-map"><span class="std std-ref">Task 3 - Map reads to the reference genome</span></a></p>
</div>
</div>
</div>
<div class="section" id="modifications-to-support-joint-population-variant-calling">
<h2>Modifications to support joint population variant calling<a class="headerlink" href="#modifications-to-support-joint-population-variant-calling" title="Permalink to this headline">¶</a></h2>
<p>This section builds on the tasks listed in and to extend variant calling
to joint calling of multiple potentially related individuals in a
population.</p>
<p>Table : Overview of population variant calling pipeline tasks</p>
<table border="1" class="docutils">
<colgroup>
<col width="5%" />
<col width="23%" />
<col width="15%" />
<col width="57%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="2">Task</th>
<th class="head">Command &amp; Utilities</th>
<th class="head">Purpose</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>3</td>
<td>Map reads to reference genome</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">map</span></code></td>
<td>Perform sex-aware mapping to the reference</td>
</tr>
<tr class="row-odd"><td>6</td>
<td>Call sequence variants</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">population</span></code></td>
<td>Perform sex-aware joint variant calls relative to the reference on a population</td>
</tr>
</tbody>
</table>
<div class="section" id="task-3-map-reads-to-the-reference-genome-population-calling">
<span id="population-map"></span><h3>Task 3 – Map reads to the reference genome (population calling)<a class="headerlink" href="#task-3-map-reads-to-the-reference-genome-population-calling" title="Permalink to this headline">¶</a></h3>
<p>For optimal results the mapping stage should be performed as described
in <a class="reference internal" href="#task-3-map-reads-to-the-reference-genome"><span class="std std-ref">Task 3 - Map reads to the reference genome</span></a>. Also ensure
that when mapping each individual they are given a
unique sample ID and that all reads for that individual have the same
sample ID (The sample ID is specified as the <code class="docutils literal"><span class="pre">SM</span></code> part of the <code class="docutils literal"><span class="pre">--sam-rg</span></code>
flag either during the formatting of the reads or for the <code class="docutils literal"><span class="pre">map</span></code> command
explicitly).</p>
<p>RTG commands for mapping and variant calling multiple samples make use
of a pedigree file specifying sample names, their sex, and any relations
between them. This is done by creating a standard PED format file
containing information about the individual samples using a text editor.
For example a PED file for the individuals NA19238, NA19239, NA19240 and
the unrelated NA12878 might look like the following.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ cat population.ped

# PED format pedigree
# fam-id ind-id pat-id mat-id sex phen
FAM01 NA19238 0 0 2 0
FAM01 NA19239 0 0 1 0
FAM01 NA19240 NA19239 NA19238 2 0
0 NA12878 0 0 2 0
</pre></div>
</div>
<p>Note that the IDs used in columns 2, 3, and 4 must match the sample IDs
used during data formatting, and the family ID in column 1 is ignored.
You can use the <code class="docutils literal"><span class="pre">pedstats</span></code> command to verify the correct format:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg pedstats population.ped
Pedigree file: population.ped
Total samples: 4
Primary samples: 4
Male samples: 1
Female samples: 3
Afflicted samples: 0
Founder samples: 3
Parent-child relationships: 2
Other relationships: 0
Families: 1
</pre></div>
</div>
<p>For more information about the PED file format see
<a class="reference internal" href="appendix.html#pedigree-ped-input-file-format"><span class="std std-ref">Pedigree PED input file format</span></a>.</p>
<p>For example when mapping the individuals NA19238, NA19239, NA19240 and
NA12878, we supply the pedigree file to automatically determine the
appropriate sex for mapping each sample. For sample NA19238:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ mkdir map_sample_NA19238
$ rtg map -i sample_NA19238/SRR002998 -t hg19 -o map_sample_NA19238/SRR002998 \
  --pedigree population.ped
</pre></div>
</div>
<p>For sample NA19239:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ mkdir map_sample_NA19239
$ rtg map -i sample_NA19239/SRR002879 -t hg19 -o map_sample_NA19239/SRR002879 \
  --pedigree population.ped
</pre></div>
</div>
<p>For sample NA19240:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ mkdir map_sample_NA19240
$ rtg map -i sample_NA19240/SRR002978 -t hg19 -o map_sample_NA19240/SRR002978 \
  --pedigree population.ped
</pre></div>
</div>
<p>And lastly for NA12878:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ mkdir map_sample_NA12878
$ rtg map -i sample_NA12878/HS2000-907_85_5_1 -t hg19 \
  -o map_sample_NA12878/HS2000-907_85_5_1 --pedigree population.ped
</pre></div>
</div>
</div>
<div class="section" id="task-6-call-population-sequence-variants">
<span id="joint-population-task6"></span><h3>Task 6 - Call population sequence variants<a class="headerlink" href="#task-6-call-population-sequence-variants" title="Permalink to this headline">¶</a></h3>
<p>The <code class="docutils literal"><span class="pre">population</span></code> command is invoked similarly to the <code class="docutils literal"><span class="pre">snp</span></code> command
except you must specify the pedigree file containing information about
each sample and any relations between them.</p>
<p>To run the <code class="docutils literal"><span class="pre">population</span></code> command on the population of NA19238, NA19239,
NA19240 and NA12878 you need to provide all the mapping files for the
samples. The mapping calibration files will be automatically detected at
the locations of the mapping files. To specify these in a file list for
input you could run:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ find /partition2/map_population -name &quot;alignments.bam&quot; &gt; map_population-bam-files
</pre></div>
</div>
<p>To run the <code class="docutils literal"><span class="pre">population</span></code> command you then specify the PED file containing
the sample ID for each member of the population in the individual ID
column.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg population --pedigree population.ped -t hg19 -o pop_variants \
  -I map_population-bam-files
</pre></div>
</div>
<p>Examining the <code class="docutils literal"><span class="pre">snps.vcf.gz</span></code> file in the output directory will show
individual columns for the variants of each member of the population.
For more details about the VCF output file see
<a class="reference internal" href="appendix.html#vcf-output-file-description"><span class="std std-ref">VCF output file description</span></a>.</p>
</div>
</div>
<div class="section" id="create-and-use-population-priors-in-variant-calling">
<h2>Create and use population priors in variant calling<a class="headerlink" href="#create-and-use-population-priors-in-variant-calling" title="Permalink to this headline">¶</a></h2>
<p>To improve the accuracy of variant calling on new members of a
population, a file containing the allele counts of the population’s
known variants may be supplied. This information is used as an extra set
of prior probabilities when making calls.</p>
<p>Sources of this allele count data can be external, for instance the 1000
genomes project, or from prior variant calling on other members of the
population. An example use case of the latter follows.</p>
<p><strong>Data</strong></p>
<p>For this use case it is assumed that the following data is available:</p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">/data/runs/20humans.vcf.gz</span></code> - output from a previous population
command run on 20 humans from a population.</li>
<li><code class="docutils literal"><span class="pre">/data/reference/human_reference</span></code> - SDF containing the human
reference sequences.</li>
<li><code class="docutils literal"><span class="pre">/data/mappings/new_human.txt</span></code> - text file containing a list of BAM
files with the sequence alignments for the new member of the
population.</li>
</ul>
<p>Table : Overview of pipeline tasks</p>
<table border="1" class="docutils">
<colgroup>
<col width="4%" />
<col width="24%" />
<col width="13%" />
<col width="58%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="2">Task</th>
<th class="head">Command &amp; Utilities</th>
<th class="head">Purpose</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>1</td>
<td>Produce population priors file</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">vcfannotate</span></code>
<code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">vcfsubset</span></code></td>
<td>Produce a reusable set of population priors from an existing VCF file</td>
</tr>
<tr class="row-odd"><td>2</td>
<td>Run variant calling using population priors</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">snp</span></code></td>
<td>Perform variant calling on the new member of the population using the new population priors to improve results</td>
</tr>
</tbody>
</table>
<div class="section" id="task-1-produce-population-priors-file">
<h3>Task 1 - Produce population priors file<a class="headerlink" href="#task-1-produce-population-priors-file" title="Permalink to this headline">¶</a></h3>
<p>Using a full VCF file for a large population as population priors can be
slow, as it contains lots of unnecessary information. The <code class="docutils literal"><span class="pre">AC</span></code> and <code class="docutils literal"><span class="pre">AN</span></code>
fields are the standard VCF specification fields representing the allele
count in genotypes, and total number of alleles in called genotypes. For
more information on these fields, see the VCF specification at
<a class="reference external" href="https://samtools.github.io/hts-specs/VCFv4.1.pdf">https://samtools.github.io/hts-specs/VCFv4.1.pdf</a>. Alternatively,
retaining only the <code class="docutils literal"><span class="pre">GT</span></code> field for each sample is sufficient, however
this is less efficient both computationally and size-wise.</p>
<p>To calculate and annotate the <code class="docutils literal"><span class="pre">AC</span></code> and <code class="docutils literal"><span class="pre">AN</span></code> fields for a VCF file, use
the RTG command <code class="docutils literal"><span class="pre">vcfannotate</span></code> with the parameter <code class="docutils literal"><span class="pre">--fill-an-ac</span></code>:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg vcfannotate --fill-an-ac -i /data/runs/20humans.vcf.gz \
  -o 20humans_an_ac.vcf.gz
</pre></div>
</div>
<p>Then remove all unnecessary data from the file using the RTG command
<code class="docutils literal"><span class="pre">vcfsubset</span></code>:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg vcfsubset --keep-info AC,AN --remove-samples \
  -i 20humans_an_ac.vcf.gz -o 20humans_priors.vcf.gz
</pre></div>
</div>
<p>This output is block compressed and tabix indexed by default, which is
necessary for the population priors input. There will be an additional
file output called <code class="docutils literal"><span class="pre">20humans_priors.vcf.gz.tbi</span></code> which is the tabix
index file.</p>
<p>The resulting population priors can now be stored in a suitable location
to be used for any further runs as required.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ cp 20humans_priors.vcf.gz* /data/population_priors/
</pre></div>
</div>
</div>
<div class="section" id="task-2-run-variant-calling-using-population-priors">
<h3>Task 2 - Run variant calling using population priors<a class="headerlink" href="#task-2-run-variant-calling-using-population-priors" title="Permalink to this headline">¶</a></h3>
<p>The population priors can now be used to improve variant calling on new
members of the population supplying the <code class="docutils literal"><span class="pre">--population-priors</span></code> parameter
to any of the variant caller commands.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg snp -o new_human_snps -t /data/reference/human_reference \
  -I /data/mappings/new_human.txt \
  --population-priors /data/population_priors/20humans_priors.vcf.gz
</pre></div>
</div>
</div>
</div>
<div class="section" id="somatic-variant-detection-in-cancer">
<h2>Somatic variant detection in cancer<a class="headerlink" href="#somatic-variant-detection-in-cancer" title="Permalink to this headline">¶</a></h2>
<p>Use the following ordered steps to detect somatic variations between
normal and tumor read samples.</p>
<p>Table : Overview of somatic pipeline tasks</p>
<table border="1" class="docutils">
<colgroup>
<col width="4%" />
<col width="23%" />
<col width="11%" />
<col width="61%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="2">Task</th>
<th class="head">Command &amp; Utilities</th>
<th class="head">Purpose</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>1</td>
<td>Format reference data</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">format</span></code></td>
<td>Convert reference sequence from FASTA file to RTG Sequence Data Format (SDF)</td>
</tr>
<tr class="row-odd"><td>2</td>
<td>Format read data</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">format</span></code></td>
<td>Convert read sequence from FASTA and FASTQ files to RTG Sequence Data Format (SDF)</td>
</tr>
<tr class="row-even"><td>3</td>
<td>Map reads against the reference genome</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">map</span></code></td>
<td>Generate read alignments for the normal and cancer samples, and report in a BAM file for downstream analysis</td>
</tr>
<tr class="row-odd"><td>4</td>
<td>Call somatic variants</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">somatic</span></code></td>
<td>Detect somatic variants between the normal and tumor samples</td>
</tr>
</tbody>
</table>
<div class="section" id="task-1-format-reference-data-somatic">
<h3>Task 1 - Format reference data (Somatic)<a class="headerlink" href="#task-1-format-reference-data-somatic" title="Permalink to this headline">¶</a></h3>
<p>Format the human reference data to RTG SDF using Task 1 of <em>RTG mapping
and sequence variant detection</em>. In the following tasks it is assumed
the human reference SDF is called <code class="docutils literal"><span class="pre">hg19</span></code>.</p>
</div>
<div class="section" id="task-2-format-read-data-somatic">
<h3>Task 2 - Format read data (Somatic)<a class="headerlink" href="#task-2-format-read-data-somatic" title="Permalink to this headline">¶</a></h3>
<p>Format the normal and tumor sample read data sets to RTG SDF using Task
2 of <em>RTG mapping and sequence variant detection</em>. In this example we
assume there are 20 lanes of data for each sample.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ for ((i=1;i&lt;20;i++)); do
$   rtg format -f fastq -q solexa -o normal_reads_${i} \
    -l /data/reads/normal/${i}/reads_1.fq \
    -r /data/reads/normal/${i}/reads_2.fq \
    --sam-rg &quot;@RG\tID:normal_${lane_id}\tSM:sm_normal\tPL:ILLUMINA&quot;
$ done
$ for ((i=1;i&lt;20;i++)); do
$   rtg format -f fastq -q solexa -o tumor_reads_${i} \
    -l /data/reads/tumor/${i}/reads_1.fq \
    -r /data/reads/tumor/${i}/reads_2.fq \
    --sam-rg &quot;@RG\tID:tumor_${lane_id}\tSM:sm_tumor\tPL:ILLUMINA&quot;
$ done
</pre></div>
</div>
</div>
<div class="section" id="task-3-map-tumor-and-normal-sample-reads-against-the-reference-genome">
<h3>Task 3 - Map tumor and normal sample reads against the reference genome<a class="headerlink" href="#task-3-map-tumor-and-normal-sample-reads-against-the-reference-genome" title="Permalink to this headline">¶</a></h3>
<p>Map the normal and tumor sample reads against the reference genome as in
<a class="reference internal" href="#task-3-map-reads-to-the-reference-genome"><span class="std std-ref">Task 3 - Map reads to the reference genome</span></a>.  The mapping must be
done with appropriate read group information for each read set with the
<code class="docutils literal"><span class="pre">--sam-rg</span></code> flag. All mappings on the normal should have the same
sample ID and all mappings on the tumor should have the same sample ID,
but the sample ID should be different between the tumor mappings and
normal mappings.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ for ((i=1;i&lt;20;i++)); do
$ rtg map -i normal_reads_${i} -t hg19 -o normal_map_${i}
$ done
$ for ((i=1;i&lt;20;i++)); do
$ rtg map -i cancer_reads_${i} -t hg19 -o cancer_map_${i}
$ done
</pre></div>
</div>
</div>
<div class="section" id="task-4-call-somatic-variants">
<h3>Task 4 - Call somatic variants<a class="headerlink" href="#task-4-call-somatic-variants" title="Permalink to this headline">¶</a></h3>
<p>The <code class="docutils literal"><span class="pre">somatic</span></code> command is invoked similarly to the <code class="docutils literal"><span class="pre">snp</span></code> command except
you need to specify some extra details. Firstly you need to specify the
sample IDs corresponding to the normal and cancer samples, with the
<code class="docutils literal"><span class="pre">--original</span></code> and <code class="docutils literal"><span class="pre">--derived</span></code> flags respectively. Secondly you may
optionally specify the estimated level of contamination from the normal
sample in the tumor sample using the <code class="docutils literal"><span class="pre">--contamination</span></code> flag.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg somatic -t hg19 -o somatic_out --derived sm_tumor --original sm_normal \
  --contamination 0.1 normal_map_*/alignments.bam tumor_map_*/alignments.bam
</pre></div>
</div>
<p>Examining the <code class="docutils literal"><span class="pre">snps.vcf.gz</span></code> file in the output directory will show a
column each for the variants of the normal and tumor samples and will
contain variants where the tumor sample differs from the normal sample.
The somatic command stores information in the VCF <code class="docutils literal"><span class="pre">INFO</span></code> fields <code class="docutils literal"><span class="pre">NCS</span></code>,
and <code class="docutils literal"><span class="pre">LOH</span></code> and <code class="docutils literal"><span class="pre">FORMAT</span></code> field <code class="docutils literal"><span class="pre">SSC</span></code> and <code class="docutils literal"><span class="pre">SS</span></code>. For more details
about the VCF output file see <a class="reference internal" href="appendix.html#vcf-output-file-description"><span class="std std-ref">VCF output file description</span></a>.</p>
</div>
<div class="section" id="using-site-specific-somatic-priors">
<h3>Using site-specific somatic priors<a class="headerlink" href="#using-site-specific-somatic-priors" title="Permalink to this headline">¶</a></h3>
<p>The <code class="docutils literal"><span class="pre">somatic</span></code> command has a default prior of 0.000001 (1e-6) for a
particular site being somatic. Since the human genome comprises some 3.2
GB, this prior corresponds to an expectation of about 3200 somatic
variants in a whole genome sample. Depending on the expected number of
variants for a particular sample, it may be appropriate to raise or
lower this prior. In general, decreasing the prior increases specificity
while increasing the prior increases sensitivity.</p>
<p>Of course, not every site is equally likely to lead to a somatic
variant. To support different priors for different sites we provide a
facility to set a prior on per site basis via a BED file we call
<code class="docutils literal"><span class="pre">site-specific</span> <span class="pre">somatic</span> <span class="pre">priors</span></code>. The <code class="docutils literal"><span class="pre">--somatic-priors</span></code> command line
option is used to provide the file to the <code class="docutils literal"><span class="pre">somatic</span></code> command.</p>
<p>The site-specific somatic priors can cover as many or as few sites as
desired. Any site not covered by a specific prior will use the default
prior. The format of the file is a BED file where the fourth column of
each line gives the explicit prior for the indicated region, for
example,</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>1 14906 14910 1e-8
</pre></div>
</div>
<p>denotes that the prior for bases 14907, 14908, 14909, and 14910 on
chromosome 1 is 1e-8 rather than the default. (Recall that BED files use
0-based indices.) If the BED file contains more than one prior covering
a particular site, then the largest prior covering that site is used.
When making a complex call, the prior used is the arithmetic average of
priors in the region of the complex call.</p>
<p>A typical starting point for making somatic site-specific priors might
include a database of known cancer sites (for example,
<a class="reference external" href="http://cancer.sanger.ac.uk/cosmic">COSMIC</a>) and a database of sites
known to be variant in the population (for example,
<a class="reference external" href="http://www.ncbi.nlm.nih.gov/SNP/">dbSNP</a>). The idea is that the
COSMIC sites are more likely to be somatic and should have a higher
prior, while those in dbSNP are less likely to be somatic and should
have a lower prior.</p>
<p>The following recipe can be used to build the BED file where some sites
have a lower prior of 1e-8 and others have a higher prior of 1e-5. The
procedure can be easily modified to incorporate additional inputs each
with its own prior.</p>
<p>First, collect prerequisites in the form of VCF files (here using the
names <code class="docutils literal"><span class="pre">cosmic.vcf.gz</span></code> and <code class="docutils literal"><span class="pre">dbsnp.vcf.gz</span></code>, but, of course, any other VCFs
can also be used).</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ COSMIC=cosmic.vcf.gz
$ DBSNP=dbsnp.vcf.gz
</pre></div>
</div>
<p>Convert each VCF file into a BED file with the desired priors taking
care to convert from 1-based coordinates in VCF to 0-based coordinates
in VCF.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ zcat ${COSMIC} | awk -vOFS=&#39;\t&#39; &#39;/^[^#]/{print $1,$2-1,$2+length($4)-1,&quot;1e-5&quot;}&#39;\
  | sort -Vu &gt;p0.bed
$ zcat ${DBSNP} | awk -vOFS=&#39;\t&#39; &#39;/^[^#]/{print $1,$2-1,$2+length($4)-1,&quot;1e-8&quot;}&#39;\
  | sort -Vu &gt;p1.bed
</pre></div>
</div>
<p>[Optional] Collapse adjacent intervals together. One way of doing this
is to use the <code class="docutils literal"><span class="pre">bedtools</span></code> <code class="docutils literal"><span class="pre">merge</span></code> facility. This can result in a smaller
final result when the intervals are dense.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ bedtools merge -c 4 -o distinct -i p0.bed &gt;p0.tmp &amp;&amp; mv p0.tmp p0.bed
$ bedtools merge -c 4 -o distinct -i p1.bed &gt;p1.tmp &amp;&amp; mv p1.tmp p1.bed
</pre></div>
</div>
<p>In general, care must be taken to ensure intersecting sites are handled
in the desired manner. Since in this case we want to use COSMIC in
preference to dbSNP and prior(COSMIC) &gt; prior(dbSNP), we can simply
merge the outputs because the somatic caller will choose the larger
prior in the case of overlap.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ sort --merge -V p0.bed p1.bed | rtg bgzip - &gt;somatic-priors.bed.gz
</pre></div>
</div>
<p>To support somatic calling on restricted regions, construct a tabix
index for the priors file.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg index -f bed somatic-priors.bed.gz
</pre></div>
</div>
<p>The site-specific somatic BED file is now ready to be used by the
<code class="docutils literal"><span class="pre">somatic</span></code> command:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg somatic -–somatic-priors somatic-priors.bed.gz ...
</pre></div>
</div>
</div>
</div>
<div class="section" id="avr-scoring-using-hapmap-for-model-building">
<h2>AVR scoring using HAPMAP for model building<a class="headerlink" href="#avr-scoring-using-hapmap-for-model-building" title="Permalink to this headline">¶</a></h2>
<p>AVR (Adaptive Variant Rescoring) is a machine learning technology for
learning and predicting which calls are likely correct. It comprises of
a learning algorithm that takes training examples and infers a model
about what constitutes a good call and a prediction engine which applies
the model to variants and estimates their correctness. It uses
attributes of the call that are not considered by the internal Bayesian
statistics model to make better predictions as to the correctness of a
variant call.</p>
<p>Each of the RTG variant callers (<code class="docutils literal"><span class="pre">snp</span></code>, <code class="docutils literal"><span class="pre">family</span></code>, <code class="docutils literal"><span class="pre">population</span></code>)
automatically runs a default AVR model, producing an AVR attribute for
each sample. The model can be changed with the <code class="docutils literal"><span class="pre">--avr-model</span></code> parameter,
and the AVR functionality can be turned off completely by specifying the
special &#8216;<code class="docutils literal"><span class="pre">none</span></code>&#8216; model.</p>
<p>Example command line usage. Turn AVR rescoring off:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg family --mother NA19238 --father NA19239 --daughter NA19240 -t hg19 \
  -o trio_variants -I map_trio-bam-files --avr-model none
</pre></div>
</div>
<p>Apply default RTG AVR model:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg family --mother NA19238 --father NA19239 --daughter NA19240 -t hg19 \
  -o trio_variants -I map_trio-bam-files
</pre></div>
</div>
<p>Apply a custom AVR model:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg family --mother NA19238 --father NA19239 --daughter NA19240 -t hg19 \
  -o trio_variants -I map_trio-bam-files --avr-model /path/to/my/custom.avr
</pre></div>
</div>
<p>The effectiveness of AVR is strongly dependent on the quality of the
training data. In general, the more training data you have, the better
the model will be. Ideally the training data should have the same
characteristics as the calls to be predicted; that is, the same
platform, the same reference, the same coverage, etc. There also needs
to be a balance of positive and negative training examples. In reality,
these conditions can only be met to varying degrees, but AVR will try to
make the most of what it is given.</p>
<p>A given AVR model is tied to a set of attributes corresponding to fields
in VCF records or quantities that are derivable from those fields. The
attributes chosen can take into account anomalies associated with
different sequencing technologies. Examples of attributes are things
like quality of the call, zygosity of the call, strand bias, allele
balance, and whether or not the call is complex. Not all attributes are
equally predictive and it is the job of the machine learning to
determine which combinations of attributes lead to the best predictions.
When building a model it is necessary to provide the list of attributes
to be used. In general, providing more attributes gives the AVR model a
better chance at learning what constitutes a good call. There are two
caveats; the attributes used during training need to be present in the
calls to be predicted and some attributes like DP are vulnerable to
changes in coverage. AVR is able to cope with missing values during both
training and prediction.</p>
<p>The training data needs to comprise both positive and negative examples.
Ideally we would know exactly the truth status of each training example,
but in reality this must be approximated by reference to some
combination of baseline information.</p>
<p>In the example that follows, the HAPMAP database will be used to produce
and then use an AVR model on a set of output variants, a process that
can be used when no appropriate AVR model is already available. The
HAPMAP database will be used to determine which of the variants will be
considered correct for training purposes. This will introduce two types
of error; correct calls which are not in HAPMAP will be marked as
negative training examples and a few incorrect calls occurring at HAPMAP
sites will be marked as positive training examples.</p>
<p><strong>Data</strong></p>
<p>Reference SDF on which variant calling was performed, in this example
assumed to be an existing SDF containing the 1000 genomes build 37 phase
2 reference</p>
<p>(<a class="reference external" href="ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/phase2_reference_assembly_sequence/hs37d5.fa.gz">ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/phase2_reference_assembly_sequence/hs37d5.fa.gz</a>).
For this example this will be called
<code class="docutils literal"><span class="pre">/data/reference/1000g_v37_phase2</span></code>.</p>
<p>The HAPMAP variants file from the Broad Institute data bundle</p>
<p>(<a class="reference external" href="ftp://gsapubftp-anonymous&#64;ftp.broadinstitute.org/bundle/2.5/b37/hapmap_3.3.b37.vcf.gz">ftp://gsapubftp-anonymous&#64;ftp.broadinstitute.org/bundle/2.5/b37/hapmap_3.3.b37.vcf.gz</a>).
For this example this will be called <code class="docutils literal"><span class="pre">/data/hapmap_3.3.b37.vcf.gz</span></code>.</p>
<p>The example will be performed on the merged results of the RTG <code class="docutils literal"><span class="pre">family</span></code>
command for the 1000 genomes CEU trio NA12878, NA12891 and NA12892. For
this example this will be called <code class="docutils literal"><span class="pre">/data/runs/NA12878trio/family.vcf.gz</span></code>.</p>
<p>Table : Overview of basic pipeline tasks</p>
<table border="1" class="docutils">
<colgroup>
<col width="5%" />
<col width="19%" />
<col width="14%" />
<col width="61%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="2">Task</th>
<th class="head">Command &amp; Utilities</th>
<th class="head">Purpose</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>1</td>
<td>Create training data</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">vcffilter</span></code></td>
<td>To generate positive and negative examples for the AVR machine learning model to train on</td>
</tr>
<tr class="row-odd"><td>2</td>
<td>Build and check AVR model</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">avrbuild</span></code>
<code class="docutils literal"><span class="pre">rtg</span> <span class="pre">avrstats</span></code></td>
<td>To create and check an AVR model</td>
</tr>
<tr class="row-even"><td>3</td>
<td>Use AVR model</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">avrpredict</span></code>
<code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">snp</span></code>
<code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">family</span></code>
<code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">population</span></code></td>
<td>To apply the AVR model to the existing output or to use it directly during variant calling</td>
</tr>
<tr class="row-odd"><td>4</td>
<td>Install AVR model</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">cp</span></code></td>
<td>Install model in standard RTG model location for later reuse</td>
</tr>
</tbody>
</table>
<div class="section" id="task-1-create-training-data">
<h3>Task 1 - Create training data<a class="headerlink" href="#task-1-create-training-data" title="Permalink to this headline">¶</a></h3>
<p>The first step is to use the <code class="docutils literal"><span class="pre">vcffilter</span></code> command to split the variant
calls into positive and negative training examples with respect to
HAPMAP. It is also possible to build training sets using the <code class="docutils literal"><span class="pre">vcfeval</span></code>
command, however this is less appropriate for dealing with sites rather
than calls, and experiments indicate using <code class="docutils literal"><span class="pre">vcffilter</span></code> leads to better
training sets.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg vcffilter --include-vcf /data/hapmap_3.3.b37.vcf.gz -o pos-NA12878.vcf.gz \
  -i /data/runs/NA12878trio/family.vcf.gz --sample NA12878 --remove-same-as-ref
$ rtg vcffilter --exclude-vcf /data/hapmap_3.3.b37.vcf.gz -o neg-NA12878.vcf.gz \
  -i /data/runs/NA12878trio/family.vcf.gz --sample NA12878 --remove-same-as-ref
</pre></div>
</div>
<p>Optionally check that the training data looks reasonable. There should
be a reasonable amount of both positive and negative examples and all
expected chromosomes should be represented.</p>
</div>
<div class="section" id="task-2-build-and-check-avr-model">
<h3>Task 2 - Build and check AVR model<a class="headerlink" href="#task-2-build-and-check-avr-model" title="Permalink to this headline">¶</a></h3>
<p>The next step is to build an AVR model. Select the attributes that will
be used with some consideration to portability and the nature of the
training set. Here we have excluded <code class="docutils literal"><span class="pre">XRX</span></code> and <code class="docutils literal"><span class="pre">LAL</span></code> because HAPMAP is
primarily SNP locations and does not capture complex calls. By excluding
<code class="docutils literal"><span class="pre">XRX</span></code> and <code class="docutils literal"><span class="pre">LAL</span></code> we prevent the model from learning that complex calls
are bad. We have also excluded <code class="docutils literal"><span class="pre">DP</span></code> because we want a model somewhat
independent of coverage level. Building the model can take a large
amount of RAM and several hours. The amount of memory required is
related to the number of training instances.</p>
<p>A list of derived annotations that can be used are available in the
documentation for <a class="reference internal" href="rtg_command_reference.html#avrbuild"><span class="std std-ref">avrbuild</span></a>. For VCF <code class="docutils literal"><span class="pre">INFO</span></code> and <code class="docutils literal"><span class="pre">FORMAT</span></code>
annotations check the header of the VCF file for available fields. The
VCF fields output by RTG variant callers are described in <a class="reference internal" href="appendix.html#vcf-output-file-description"><span class="std std-ref">VCF output file description</span></a>.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg avrbuild -o NA12878model.avr --info-annotations DPR \
  --format-annotations DPR,AR,ABP,SBP,RPB,GQ \
  --derived-annotations IC,EP,QD,AN,PD,GQD,ZY \
  --sample NA12878 -p pos-NA12878.vcf.gz -n neg-NA12878.vcf.gz
</pre></div>
</div>
<p>Examine the statistics output to the screen to check things look
reasonable. Due to how attributes are computed there can be missing
values, but it is a bad sign if any attribute is missing from most
samples.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>Total number of examples: 5073752
Total number of positive examples: 680677
Total number of negative examples: 4393075
Total weight of positive examples: 2536873.27
Total weight of negative examples: 2536876.42
Number of examples with missing values:
DERIVED-AN 0
DERIVED-EP 0
DERIVED-GQD 0
DERIVED-IC 915424
DERIVED-PD 0
DERIVED-QD 0
DERIVED-ZY 0
FORMAT-ABP 13602
FORMAT-AR 8375
FORMAT-DPR 0
FORMAT-GQ 0
FORMAT-RPB 8375
FORMAT-SBP 37129
INFO-DPR 0
Hold-out score: 69.1821% (96853407/139997752)
Attribute importance estimate for DERIVED-AN: 0.0304% (29443/96853407)
Attribute importance estimate for DERIVED-EP: 1.2691% (1229135/96853407)
Attribute importance estimate for DERIVED-GQD: 4.7844%
(4633873/96853407)
Attribute importance estimate for DERIVED-IC: 4.7748% (4624554/96853407)
Attribute importance estimate for DERIVED-PD: 0.0000% (0/96853407)
Attribute importance estimate for DERIVED-QD: 6.8396% (6624383/96853407)
Attribute importance estimate for DERIVED-ZY: 0.8693% (841925/96853407)
Attribute importance estimate for FORMAT-ABP: 5.4232% (5252533/96853407)
Attribute importance estimate for FORMAT-AR: 2.6008% (2518955/96853407)
Attribute importance estimate for FORMAT-DPR: 3.3186% (3214163/96853407)
Attribute importance estimate for FORMAT-GQ: 3.4268% (3319009/96853407)
Attribute importance estimate for FORMAT-RPB: 0.2176% (210761/96853407)
Attribute importance estimate for FORMAT-SBP: 0.0219% (21196/96853407)
Attribute importance estimate for INFO-DPR: 4.6525% (4506074/96853407)
</pre></div>
</div>
<p>Optionally check the resulting model file using the <code class="docutils literal"><span class="pre">avrstats</span></code> command.
This will produce a short summary of the model including the attributes
used during the build.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg avrstats NA12878model.avr
Location : NA12878model.avr
Parameters : avrbuild -o NA12878model.avr --info-annotations DPR --derived-annotations IC,EP,QD,AN,PD,GQD,ZY --format-annotations DPR,AR,ABP,SBP,RPB,GQ --sample NA12878 -p pos-NA12878.vcf.gz -n neg-NA12878.vcf.gz
Date built : 2013-05-17-09-41-17
AVR Version : 1
AVR-ID : 7ded37d7-817f-467b-a7da-73e374719c7f
Type : ML
QUAL used : false
INFO fields : DPR
FORMAT fields : DPR,AR,ABP,SBP,RPB,GQ
Derived fields: IC,EP,QD,AN,PD,GQD,ZY
</pre></div>
</div>
</div>
<div class="section" id="task-3-use-avr-model">
<h3>Task 3 - Use AVR model<a class="headerlink" href="#task-3-use-avr-model" title="Permalink to this headline">¶</a></h3>
<p>The model is now ready to be used. It can be applied to the existing
variant calling output by using the <code class="docutils literal"><span class="pre">avrpredict</span></code> command.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg avrpredict --avr-model NA12878model.avr \
  -i /data/runs/NA12878trio/family.vcf.gz -o predict.vcf.gz
</pre></div>
</div>
<p>This will create or update the <code class="docutils literal"><span class="pre">AVR</span></code> <code class="docutils literal"><span class="pre">FORMAT</span></code> fields in the VCF output
file with a score between 0 and 1. The higher the resulting score the
more likely it is correct. To select an appropriate cutoff value for
further analysis of variants some approaches might include measuring the
Ti/Tv ratio or measuring sensitivity against another standard such as
OMNI at varying score cutoffs.</p>
<p>The model can also be used directly in any new variant calling runs:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg snp --avr-model NA12878model.avr -t hg19 -o snp_sample_NA19240 --sex female \
  -I sample_NA19240-map-files
$ rtg population --avr-model NA12878model.avr -t hg19 -o pop_variants \
  -I map_population-bam-files
</pre></div>
</div>
</div>
<div class="section" id="task-4-install-avr-model">
<h3>Task 4 - Install AVR model<a class="headerlink" href="#task-4-install-avr-model" title="Permalink to this headline">¶</a></h3>
<p>The custom AVR model can be installed into a standard location so that
it can be referred to by a short name (rather than the full file path
name) in the <code class="docutils literal"><span class="pre">avrpredict</span></code> and variant caller commands. The default
location for AVR models is within a subdirectory of the RTG installation
directory called <code class="docutils literal"><span class="pre">models</span></code>, and each file in that directory with a <code class="docutils literal"><span class="pre">.avr</span></code>
extension is a model that can be accessed by its short name. For example
if the <code class="docutils literal"><span class="pre">NA12878model.avr</span></code> model file is placed in
<code class="docutils literal"><span class="pre">/path/to/rtg/installation/models/NA12878model.avr</span></code> it can be accessed
by any user either using the full path to the model:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg snp --avr-model /path/to/rtg/installation/models/NA12878model.avr \
  -o snp_sample_NA19240 -t hg19 --sex female -I sample_NA19240-map-files
</pre></div>
</div>
<p>or, by just the model file name:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg snp --avr-model NA12878model.avr -o snp_sample_NA19240 -t hg19 --sex female \
  -I sample_NA19240-map-files
</pre></div>
</div>
<p>The AVR model directory will already contain the models prebuilt by RTG:</p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">illumina-exome.avr</span></code> - model built from Illumina exome sequencing
data. This model is the default model when running normal variant calling.</li>
<li><code class="docutils literal"><span class="pre">illumina-wgs.avr</span></code> - model built from Illumina whole genome
sequencing data. If you are running variant calling Illumina WGS you
may want to use this model instead of the default, although the
default should still be effective.</li>
<li><code class="docutils literal"><span class="pre">illumina-somatic.avr</span></code> - model built from Illumina whole genome
sequencing data, which is applicable to somatic variant calling,
where a variety of allelic fractions are to be expected in somatic
variants. The <code class="docutils literal"><span class="pre">somatic</span></code> command defaults to this AVR model.</li>
<li><code class="docutils literal"><span class="pre">alternate.avr</span></code> - model built using <code class="docutils literal"><span class="pre">XRX</span></code>, <code class="docutils literal"><span class="pre">ZY</span></code> and <code class="docutils literal"><span class="pre">GQD</span></code> attributes.
This should be platform independent and may be a better choice if a
more specific model for your data is unavailable. In particular, this
model may be more appropriate for scoring the results of variant
calling in situations where unusual allele-balance is expected (for
example somatic calling with contamination, or calling high
amplification data where allele drop out is expected)</li>
</ul>
</div>
</div>
<div class="section" id="rtg-structural-variant-detection">
<h2>RTG structural variant detection<a class="headerlink" href="#rtg-structural-variant-detection" title="Permalink to this headline">¶</a></h2>
<p>RTG has developed tools to assist in discovering structural variant
regions in whole genome sequencing data. The tools can be used to locate
likely structural variant breakpoints and regions that have been
duplicated or deleted. These tools are capable of processing whole
genome mapping data containing multiple read groups in a streamlined
fashion.</p>
<p>Table : Overview of structural variants analysis pipeline tasks</p>
<table border="1" class="docutils">
<colgroup>
<col width="5%" />
<col width="24%" />
<col width="12%" />
<col width="60%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="2">Task</th>
<th class="head">Command &amp; Utilities</th>
<th class="head">Purpose</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>1</td>
<td>Format reference and read data</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">format</span></code></td>
<td>Convert reference and read sequence from FASTA file to RTG Sequence Data Format (SDF)</td>
</tr>
<tr class="row-odd"><td>2</td>
<td>Map reads against a reference genome</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">map</span></code></td>
<td>Generate read alignments against a given reference, and report in a BAM file for downstream analysis</td>
</tr>
<tr class="row-even"><td>3</td>
<td>Find structural variants with sv</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">sv</span></code></td>
<td>Process prepared mapping results to identify likely structural variants</td>
</tr>
<tr class="row-odd"><td>4</td>
<td>Find structural variants with discord</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">discord</span></code></td>
<td>Process prepared mapping results to identify likely structural variant breakends</td>
</tr>
</tbody>
</table>
<div class="section" id="task-1-format-reference-and-read-data">
<h3>Task 1 - Format reference and read data<a class="headerlink" href="#task-1-format-reference-and-read-data" title="Permalink to this headline">¶</a></h3>
<p>Format the human reference data and the sample read data to RTG SDF
using Tasks 1 and 2 of <em>RTG mapping and sequence variant detection</em>. In
the following tasks it is assumed the human reference SDF is called
<code class="docutils literal"><span class="pre">hg19</span></code> and the read sample SDF is called <code class="docutils literal"><span class="pre">sample_NA19240/RG0022101</span></code>.</p>
</div>
<div class="section" id="task-2-map-reads-against-a-reference-genome">
<h3>Task 2 - Map reads against a reference genome<a class="headerlink" href="#task-2-map-reads-against-a-reference-genome" title="Permalink to this headline">¶</a></h3>
<p>This mapping proceeds as normal, as described in Task 3 of <em>RTG mapping
and sequence variant detection</em>. For structural variant detection it is
particularly important to specify the read group information with the
<code class="docutils literal"><span class="pre">--sam-rg</span></code> flag either during the formatting of the reads or for the map
command explicitly. The structural variant tools currently requires the
<code class="docutils literal"><span class="pre">PL</span></code> (platform) attribute to be either <code class="docutils literal"><span class="pre">ILLUMINA</span></code> (for Illumina reads)
or <code class="docutils literal"><span class="pre">COMPLETE</span></code> (for Complete Genomics reads).</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg map -i sample_NA19240/RG0022101 -t hg19 -o map_NA19240/RG0022101 \
  --sam-rg &quot;@RG\tID:RG0022101\tSM:NA19240\tPL:ILLUMINA&quot;
</pre></div>
</div>
<p>Mapping identifies discordant read matings and inserts the pair
information for unique unmated reads into the SAM records. The <code class="docutils literal"><span class="pre">map</span></code>
command also produces a file within the directory called <code class="docutils literal"><span class="pre">rgstats.tsv</span></code>
containing read group statistics.</p>
<p>The structural variant callers require the read group statistics files
to be supplied, so as with multiple BAM files, it is possible to create
a text file listing the locations of all the required statistics files:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ find /partition2/map_sample_NA19240 -name &quot;rgstats.tsv&quot; \
  &gt; sample_NA19240-rgstats-files
</pre></div>
</div>
</div>
<div class="section" id="task-4-find-structural-variants-with-sv">
<h3>Task 4 - Find structural variants with sv<a class="headerlink" href="#task-4-find-structural-variants-with-sv" title="Permalink to this headline">¶</a></h3>
<p>Once mapping is complete one can run the structural variants analysis
tool. To run the <code class="docutils literal"><span class="pre">sv</span></code> tool you need to supply the mapping BAM files and
the read group statistics files. As with the <code class="docutils literal"><span class="pre">snp</span></code>, <code class="docutils literal"><span class="pre">cnv</span></code> and <code class="docutils literal"><span class="pre">coverage</span></code>
tools, this can be a large number of files and so input can be specified
using list files.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg sv -t hg19 -o sample_NA19240-sv -I sample_NA19240-bam-files \
  -R sample_NA19240-rgstats-files
$ ls -l sample_NA19240-sv
    30 Oct 1 16:56 done
 15440 Oct 1 16:56 progress
     0 Oct 1 16:56 summary.txt
122762 Oct 1 16:56 sv_bayesian.tsv.gz
  9032 Oct 1 16:56 sv.log
</pre></div>
</div>
<p>The file <code class="docutils literal"><span class="pre">sv_bayesian.tsv.gz</span></code> contains a trace of the strengths of
alternative bayesian hypothesis at points along the reference genome.
The currently supported hypotheses are shown in the following table.</p>
<p>Table : Structural variant hypotheses</p>
<table border="1" class="docutils">
<colgroup>
<col width="18%" />
<col width="82%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Hypothesis</th>
<th class="head">Semantics</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">normal</span></code></td>
<td>Normal mappings, no structural variants</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">duplicate</span></code></td>
<td>Above normal mappings, potential duplication region</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">delete</span></code></td>
<td>Below normal mappings, potential deletion region</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">delete-left</span></code></td>
<td>Mapping data suggest the left breakpoint of a deletion</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">delete-right</span></code></td>
<td>Mapping data suggest the right breakpoint of a deletion</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">duplicate-left</span></code></td>
<td>Mapping data suggest the left boundary of a region that has been copied elsewhere</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">duplicate-right</span></code></td>
<td>Mapping data suggest the right boundary of a region that has been copied elsewhere</td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">breakpoint</span></code></td>
<td>Mapping data suggest this location has received an insertion of copied genome</td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">novel-insertion</span></code></td>
<td>Mapping data suggests this location has received an insertion of material not present in the reference</td>
</tr>
</tbody>
</table>
<p>For convenience, the last column of the output file gives the index of
the hypothesis with the maximum strength, to make it easier to identify
regions where this changes for further investigation by the researcher.</p>
<p>The <code class="docutils literal"><span class="pre">sv</span></code> command also supports calling on individual chromosomes (or
regions within a chromosome) with the <code class="docutils literal"><span class="pre">--region</span></code> parameter, and this can
be used to increase overall throughput.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ for seq in M {1..22} X Y; do
$ rtg sv -o sv_sample_NA19240-chr${seq} -t hg19 -I sample_NA19240-sam-files \
  -R sample_NA19240-rgstats-files --region chr${seq}
$ done
</pre></div>
</div>
</div>
<div class="section" id="task-5-find-structural-variants-with-discord">
<h3>Task 5 - Find structural variants with discord<a class="headerlink" href="#task-5-find-structural-variants-with-discord" title="Permalink to this headline">¶</a></h3>
<p>A second tool for finding structural variant break-ends is based on
detecting cluster of discordantly mapped reads, those where both ends of
a read are mapped but are mapped either in an unexpected orientation or
with a TLEN outside the normal range for that read group. As with the
<code class="docutils literal"><span class="pre">sv</span></code> command, <code class="docutils literal"><span class="pre">discord</span></code> requires the read group statistics to be
supplied.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg discord -t hg19 -o discord_sample_NA19240 -I sample_NA19240-bam-files \
  -R sample_NA19240-rgstats-files --bed
</pre></div>
</div>
<p>As with <code class="docutils literal"><span class="pre">sv</span></code>, the <code class="docutils literal"><span class="pre">discord</span></code> command also supports using the <code class="docutils literal"><span class="pre">--region</span></code>
flag:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ for seq in M {1..22} X Y; do
$ rtg discord -o discord_sample_NA19240-chr${seq} -t hg19 \
  -I sample_NA19240-bam-files -R sample_NA19240-rgstats-files \
  --region chr${seq}
$ done
</pre></div>
</div>
<p>The default output is in VCF format, following the VCF 4.1 specification
for break-ends. However, as most third-party tools currently don’t
support this type of VCF, it is also possible to output each break-end
as a separate region in a BED file.</p>
</div>
</div>
<div class="section" id="ion-torrent-bacterial-mapping-and-sequence-variant-detection">
<h2>Ion Torrent bacterial mapping and sequence variant detection<a class="headerlink" href="#ion-torrent-bacterial-mapping-and-sequence-variant-detection" title="Permalink to this headline">¶</a></h2>
<p>The following example supports the steps typical to bacterial genome
analysis in which an Ion Torrent sequencer has generated reads at 10x
coverage.</p>
<p><strong>Data</strong></p>
<p>The baseline uses actual data pulled from the Ion community. The
reference sequence is &#8216;Escherichia coli K-12 sub-strain DH10B&#8217;, available
from the NCBI RefSeq database, accession NC_010473
(<a class="reference external" href="http://www.ncbi.nlm.nih.gov/nuccore/NC_010473">http://www.ncbi.nlm.nih.gov/nuccore/NC_010473</a>). The read data is
comprised of Ion Torrent PGM run B14-387, which can be found at the Ion
community (<a class="reference external" href="http://lifetech-it.hosted.jivesoftware.com">http://lifetech-it.hosted.jivesoftware.com</a>, requires
registration).</p>
<p>Table : Overview of basic pipeline tasks</p>
<table border="1" class="docutils">
<colgroup>
<col width="4%" />
<col width="23%" />
<col width="11%" />
<col width="61%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="2">Task</th>
<th class="head">Command &amp; Utilities</th>
<th class="head">Purpose</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>1</td>
<td>Format reference data</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">format</span></code></td>
<td>Convert reference sequence from FASTA file to RTG Sequence Data Format (SDF)</td>
</tr>
<tr class="row-odd"><td>2</td>
<td>Format read data</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">format</span></code></td>
<td>Convert read sequence from FASTA and FASTQ files to RTG Sequence Data Format (SDF)</td>
</tr>
<tr class="row-even"><td>3</td>
<td>Map reads against the reference genome</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">map</span></code></td>
<td>Generate read alignments for the normal and cancer samples, and report in a BAM file for downstream analysis</td>
</tr>
<tr class="row-odd"><td>4</td>
<td>Call sequence variants in haploid mode</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">snp</span></code></td>
<td>Detect SNPs, MNPs, and indels in haploid sample relative to the reference genome</td>
</tr>
</tbody>
</table>
<div class="section" id="task-1-format-bacterial-reference-data-ion-torrent">
<h3>Task 1 - Format bacterial reference data (Ion Torrent)<a class="headerlink" href="#task-1-format-bacterial-reference-data-ion-torrent" title="Permalink to this headline">¶</a></h3>
<p>Mapping and variant detection requires a conversion of the reference
genome from FASTA files into the RTG SDF format. This task will be
completed with the <code class="docutils literal"><span class="pre">format</span></code> command. The conversion will create an SDF
directory containing the reference genome.</p>
<p>Use the <code class="docutils literal"><span class="pre">format</span></code> command to convert the FASTA file into an SDF directory
for the reference genome.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg format -o ecoli-DH10B \
  /data/bacteria/Escherichia_coli_K_12_substr__DH10B_uid58979/NC_010473.fna
</pre></div>
</div>
</div>
<div class="section" id="task-2-format-read-data-ion-torrent">
<h3>Task 2 - Format read data (Ion Torrent)<a class="headerlink" href="#task-2-format-read-data-ion-torrent" title="Permalink to this headline">¶</a></h3>
<p>Mapping and variant detection of Ion Torrent data requires a conversion
of the read sequence data from FASTQ files into the RTG SDF format.
Additionally, it is recommended that read trimming based on the quality
data present within the FASTQ file be performed as part of this
conversion.</p>
<p>Use the <code class="docutils literal"><span class="pre">format</span></code> command to convert the read FASTQ file into an SDF
directory, using the quality threshold option to trim poor quality ends
of reads.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg format -f fastq -q solexa --trim-threshold=15 -o B14-387-reads \
  /data/reads/R_2011_07_19_20_05_38_user_B14-387-r121336-314_pool30-ms_Auto_B14\
-387-r121336-314_pool30-ms_8399.fastq
</pre></div>
</div>
</div>
<div class="section" id="task-3-map-ion-torrent-reads-to-the-reference-genome">
<h3>Task 3 - Map Ion Torrent reads to the reference genome<a class="headerlink" href="#task-3-map-ion-torrent-reads-to-the-reference-genome" title="Permalink to this headline">¶</a></h3>
<p>Map the sequence reads against the reference genome to generate
alignments in BAM format.</p>
<p>The RTG <code class="docutils literal"><span class="pre">map</span></code> command provides a means for the mapping of Ion Torrent
reads to a reference genome. When mapping Ion Torrent reads, a read
group with the platform field (<code class="docutils literal"><span class="pre">PL</span></code>) set to <code class="docutils literal"><span class="pre">IONTORRENT</span></code> should be
provided.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg map -i B14-387-reads -t ecoli-DH10B -o B14-387-map \
  --sam-rg &quot;@RG\tID:B14-387-r121336\tSM:B14-387\tPL:IONTORRENT&quot;
</pre></div>
</div>
<p>Multiple files are written to the output directory of the mapping run.
For further variant calling, the <code class="docutils literal"><span class="pre">alignments.bam</span></code> file has the
associated required index file <code class="docutils literal"><span class="pre">alignments.bam.bai</span></code>. The additional
files <code class="docutils literal"><span class="pre">alignments.bam.calibration</span></code> contains metadata to provide more
accurate variant calling.</p>
</div>
<div class="section" id="task-4-call-sequence-variants-in-haploid-mode">
<h3>Task 4 - Call sequence variants in haploid mode<a class="headerlink" href="#task-4-call-sequence-variants-in-haploid-mode" title="Permalink to this headline">¶</a></h3>
<p>Call haploid sequence variants in the mapped reads against the reference
genome to generate a variants file in the VCF format.</p>
<p>The <code class="docutils literal"><span class="pre">snp</span></code> command will automatically set machine error calibration
parameters according to the platform (PL attribute) specified in the SAM
read group, in this example to the Ion Torrent parameters. The <code class="docutils literal"><span class="pre">snp</span></code>
command defaults to diploid variant calling, so for this bacterial
example haploid mode will be specified. The automatically included
<code class="docutils literal"><span class="pre">.calibration</span></code> files provide additional information specific to the
mapping data for improved variant calling.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg snp -t ecoli-DH10B -o B14-387-snp --ploidy=haploid B14-387-map/alignments.bam
</pre></div>
</div>
<p>Examining the <code class="docutils literal"><span class="pre">snps.vcf.gz</span></code> file in the output directory will show that
variants have been called in haploid mode. For more details about the
VCF output file see <a class="reference internal" href="appendix.html#vcf-output-file-description"><span class="std std-ref">VCF output file description</span></a>.</p>
</div>
</div>
<div class="section" id="rtg-contaminant-filtering">
<h2>RTG contaminant filtering<a class="headerlink" href="#rtg-contaminant-filtering" title="Permalink to this headline">¶</a></h2>
<p>Use the following set of tasks to remove contaminated reads from a
sequenced DNA sample.</p>
<p>The RTG contamination filter, called <code class="docutils literal"><span class="pre">mapf</span></code>, removes contaminant reads
by mapping against a database of potential contaminant sequences. For
example, a bacterial metagenomic sample may have some amount of human
sequence contaminating it. The following process removes any human reads
leaving only bacteria reads.</p>
<p>Table : Overview of contaminant filtering tasks</p>
<table border="1" class="docutils">
<colgroup>
<col width="5%" />
<col width="18%" />
<col width="14%" />
<col width="63%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="2">Task</th>
<th class="head">Command &amp; Utilities</th>
<th class="head">Purpose</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>1</td>
<td>Format reference data</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">format</span></code></td>
<td>Convert reference sequence from FASTA file to RTG Sequence Data Format (SDF)</td>
</tr>
<tr class="row-odd"><td>2</td>
<td>Format read data</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">format</span></code></td>
<td>Convert read sequence from FASTA and FASTQ files to RTG Sequence Data Format (SDF)</td>
</tr>
<tr class="row-even"><td>3</td>
<td>Run contamination filter</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">mapf</span></code></td>
<td>Produce the SDF file of reads which map to the contaminant and the SDF file of those that do not</td>
</tr>
<tr class="row-odd"><td>4</td>
<td>Manage filtered reads</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">mv</span></code></td>
<td>Set up the results for use in further processing</td>
</tr>
</tbody>
</table>
<div class="section" id="task-1-format-reference-data-contaminant-filtering">
<h3>Task 1 - Format reference data (contaminant filtering)<a class="headerlink" href="#task-1-format-reference-data-contaminant-filtering" title="Permalink to this headline">¶</a></h3>
<p>RTG tools require a conversion of reference sequences from FASTA files
into the RTG SDF format. This task will be completed with the <code class="docutils literal"><span class="pre">format</span></code>
command. The conversion will create an SDF directory containing the
reference genome.</p>
<p>First, observe a typical genome reference with multiple chromosome
sequences stored in compressed FASTA format.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ ls -l /data/human/hg19/
43026389 Mar 21 2009 chr10.fa.gz
42966674 Mar 21 2009 chr11.fa.gz
42648875 Mar 21 2009 chr12.fa.gz
31517348 Mar 21 2009 chr13.fa.gz
28970334 Mar 21 2009 chr14.fa.gz
26828094 Mar 21 2009 chr15.fa.gz
25667827 Mar 21 2009 chr16.fa.gz
25139792 Mar 21 2009 chr17.fa.gz
24574571 Mar 21 2009 chr18.fa.gz
17606811 Mar 21 2009 chr19.fa.gz
73773666 Mar 21 2009 chr1.fa.gz
19513342 Mar 21 2009 chr20.fa.gz
11549785 Mar 21 2009 chr21.fa.gz
11327826 Mar 21 2009 chr22.fa.gz
78240395 Mar 21 2009 chr2.fa.gz
64033758 Mar 21 2009 chr3.fa.gz
61700369 Mar 21 2009 chr4.fa.gz
58378199 Mar 21 2009 chr5.fa.gz
54997756 Mar 21 2009 chr6.fa.gz
50667196 Mar 21 2009 chr7.fa.gz
46889258 Mar 21 2009 chr8.fa.gz
39464200 Mar 21 2009 chr9.fa.gz
    5537 Mar 21 2009 chrM.fa.gz
49278128 Mar 21 2009 chrX.fa.gz
 8276338 Mar 21 2009 chrY.fa.gz
</pre></div>
</div>
<p>Now, use the <code class="docutils literal"><span class="pre">format</span></code> command to convert multiple input files into a
single SDF directory for the reference.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg format -o hg19 /data/human/hg19/chrM.fa.gz \
  /data/human/hg19/chr[1-9].fa.gz \
  /data/human/hg19/chr1[0-9].fa.gz \
  /data/human/hg19/chr2[0-9].fa.gz \
  /data/human/hg19/chrX.fa.gz \
  /data/human/hg19/chrY.fa.gz
</pre></div>
</div>
<p>This takes the human reference FASTA files and creates a directory
called <code class="docutils literal"><span class="pre">hg19</span></code> containing the SDF, with chromosomes ordered in the
standard UCSC ordering. You can use the <code class="docutils literal"><span class="pre">sdfstats</span></code> command to show
statistics for your reference SDF, including the individual sequence
lengths.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg sdfstats --lengths hg19
Type : DNA
Number of sequences: 25
Maximum length : 249250621
Minimum length : 16571
Sequence names : yes
N : 234350281
A : 844868045
C : 585017944
G : 585360436
T : 846097277
Total residues : 3095693983
Residue qualities : no
Sequence lengths:
chrM 16571
chr1 249250621
chr2 243199373
chr3 198022430
chr4 191154276
chr5 180915260
chr6 171115067
chr7 159138663
chr8 146364022
chr9 141213431
chr10 135534747
chr11 135006516
chr12 133851895
chr13 115169878
chr14 107349540
chr15 102531392
chr16 90354753
chr17 81195210
chr18 78077248
chr19 59128983
chr20 63025520
chr21 48129895
chr22 51304566
chrX 155270560
chrY 59373566
</pre></div>
</div>
</div>
<div class="section" id="task-2-format-read-data-contaminant-filtering">
<h3>Task 2 - Format read data (contaminant filtering)<a class="headerlink" href="#task-2-format-read-data-contaminant-filtering" title="Permalink to this headline">¶</a></h3>
<p>RTG tools require a conversion of read sequence data from FASTA or FASTQ
files into the RTG SDF format. This task will be completed with the
<code class="docutils literal"><span class="pre">format</span></code> command. The conversion will create an SDF directory for the
sample reads.</p>
<p>Take a paired set of reads in FASTQ format and convert it into RTG data
format (SDF). This example shows one run of data, taking as input both
left and right mate pairs from the same run.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg format -f fastq -q sanger -o bacteria-sample \
  -l /data/reads/bacteria/sample_1.fq \
  -r /data/reads/bacteria/sample_2.fq
</pre></div>
</div>
<p>This creates a directory named bacteria-sample with two subdirectories,
named left and right. Use the <code class="docutils literal"><span class="pre">sdfstats</span></code> command to verify this step.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg sdfstats bacteria-sample
</pre></div>
</div>
</div>
<div class="section" id="task-3-run-contamination-filter">
<h3>Task 3 - Run contamination filter<a class="headerlink" href="#task-3-run-contamination-filter" title="Permalink to this headline">¶</a></h3>
<p>The <code class="docutils literal"><span class="pre">mapf</span></code> command functions in much the same way as the <code class="docutils literal"><span class="pre">map</span></code> command,
but instead of producing BAM files of the mappings it produces two SDF
directories, one containing reads that map to the reference and the
other with reads that do not map. As with the <code class="docutils literal"><span class="pre">map</span></code> command there are
multiple tuning parameters to adjust sensitivity and selectivity of the
mappings. As with the <code class="docutils literal"><span class="pre">map</span></code> command, you can use the <code class="docutils literal"><span class="pre">--start-read</span></code> and
<code class="docutils literal"><span class="pre">--end-read</span></code> flags to perform the mapping in smaller sections if
required. The default <code class="docutils literal"><span class="pre">mapf</span></code> settings are similar to <code class="docutils literal"><span class="pre">map</span></code> although the
word size and step sizes have been adjusted to yield more sensitive
mappings.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg mapf -t hg19 -i bacteria-sample -o filter-sample
</pre></div>
</div>
<p>In the <code class="docutils literal"><span class="pre">filter-sample</span></code> output directory there are, amongst other files,
two directories named <code class="docutils literal"><span class="pre">alignments.sdf</span></code> and <code class="docutils literal"><span class="pre">unmapped.sdf</span></code>. The
<code class="docutils literal"><span class="pre">alignments.sdf</span></code> directory is an SDF of the reads that mapped to the
reference, and the <code class="docutils literal"><span class="pre">unmapped.sdf</span></code> directory is an SDF of the reads that
did not map.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ ls -l filter-sample/
   4096 Sep 30 16:02 alignments.sdf/
     33 Sep 30 16:02 done
2776886 Sep 30 16:02 mapf.log
  12625 Sep 30 16:02 progress
    143 Sep 30 16:02 summary.txt
   4096 Sep 30 16:02 unmapped.sdf/
</pre></div>
</div>
</div>
<div class="section" id="task-4-manage-filtered-reads">
<h3>Task 4 - Manage filtered reads<a class="headerlink" href="#task-4-manage-filtered-reads" title="Permalink to this headline">¶</a></h3>
<p>Depending on the use case, either rename, move or delete the filtered
SDF directories as necessary. In this example the reads that did not map
to the contamination reference are to be used in further processing, so
rename the <code class="docutils literal"><span class="pre">unmapped.sdf</span></code> directory.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ mv filter-sample/unmapped.sdf bacteria-sample-filtered
</pre></div>
</div>
<p>The filtered read set is now ready for subsequent processing, such as
with the <code class="docutils literal"><span class="pre">mapx</span></code> or <code class="docutils literal"><span class="pre">species</span></code> tools.</p>
</div>
</div>
<div class="section" id="rtg-translated-protein-searching">
<h2>RTG translated protein searching<a class="headerlink" href="#rtg-translated-protein-searching" title="Permalink to this headline">¶</a></h2>
<p>Use the following set of tasks to search DNA reads against a protein
data set.</p>
<p>The RTG protein search tool, <code class="docutils literal"><span class="pre">mapx</span></code> translates nucleotide reads into
protein space and search them against a protein data set. For example, a
sample taken from a human gut can be searched against a protein data set
to determine which kinds of protein families are present in the sample.</p>
<p>For this example we will search a human gut sample read set against an
NCBI non-redundant protein data set. In the following tasks it is
assumed non-redundant protein data set is called <code class="docutils literal"><span class="pre">nr.fasta</span></code> and the
human gut sample is called <code class="docutils literal"><span class="pre">human-gut.fastq</span></code>.</p>
<p>Table : Overview of translated protein searching tasks</p>
<table border="1" class="docutils">
<colgroup>
<col width="5%" />
<col width="23%" />
<col width="14%" />
<col width="57%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="2">Task</th>
<th class="head">Command &amp; Utilities</th>
<th class="head">Purpose</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>1</td>
<td>Format protein data set</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">format</span></code></td>
<td>Convert protein data set from FASTA to RTG sequence data format (SDF)</td>
</tr>
<tr class="row-odd"><td>2</td>
<td>Format DNA read set</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">format</span></code></td>
<td>Convert read sequence from FASTA and FASTQ files to RTG Sequence Data Format (SDF)</td>
</tr>
<tr class="row-even"><td>3</td>
<td>Search against protein data set</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">mapx</span></code></td>
<td>Generate search results with alignments in tabular format</td>
</tr>
</tbody>
</table>
<div class="section" id="task-1-format-protein-data-set">
<h3>Task 1 - Format protein data set<a class="headerlink" href="#task-1-format-protein-data-set" title="Permalink to this headline">¶</a></h3>
<p>The <code class="docutils literal"><span class="pre">mapx</span></code> command requires a conversion of a protein data set from
FASTA files into RTG SDF format. This task will be completed with the
<code class="docutils literal"><span class="pre">format</span></code> command. The conversion will create an SDF directory containing
the protein data set.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg format -p -o nr /data/NCBI-nr/nr.fasta
</pre></div>
</div>
<p>The above command will take the <code class="docutils literal"><span class="pre">nr.fasta</span></code> file and create a directory
called <code class="docutils literal"><span class="pre">nr</span></code> containing the SDF. Note that the <code class="docutils literal"><span class="pre">-p</span></code> option is used to
create the SDF with protein data.</p>
</div>
<div class="section" id="task-2-format-dna-read-set">
<h3>Task 2 - Format DNA read set<a class="headerlink" href="#task-2-format-dna-read-set" title="Permalink to this headline">¶</a></h3>
<p>The <code class="docutils literal"><span class="pre">mapx</span></code> command requires a conversion of the DNA read set data from
FASTA or FASTQ files into RTG SDF format. This task will be completed
with the <code class="docutils literal"><span class="pre">format</span></code> command. The following command assumes the sample read
data set is in Solexa FASTQ format.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg format -f fastq -q solexa -o human-gut /data/human-gut-sample.fastq
</pre></div>
</div>
</div>
<div class="section" id="task-3-search-against-protein-data-set">
<h3>Task 3 - Search against protein data set<a class="headerlink" href="#task-3-search-against-protein-data-set" title="Permalink to this headline">¶</a></h3>
<p>Search the DNA reads against the protein data set and generate
alignments in tabular format.</p>
<p>The <code class="docutils literal"><span class="pre">mapx</span></code> command provides multiple tuning parameters to adjust
sensitivity and selectivity at the search stage. As with the <code class="docutils literal"><span class="pre">map</span></code>
command, you can use the <code class="docutils literal"><span class="pre">--start-read</span></code> and <code class="docutils literal"><span class="pre">--end-read</span></code> flags to
perform the mapping in smaller sections if required. In general, protein
search strategies are based on protein similarity also known as
identity.</p>
<p>The search example below uses a sensitivity setting that will guarantee
reporting with reads that align with 4 substitutions and 1 indels.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg mapx -t nr -i human-gut -o mapx_results -a 3 -b 1
</pre></div>
</div>
<p>The <code class="docutils literal"><span class="pre">alignments.tsv.gz</span></code> file in the <code class="docutils literal"><span class="pre">mapx_results</span></code> output directory
contains tabular output with alignments. For more information about this
output format see <a class="reference internal" href="appendix.html#mapx-output-file-description"><span class="std std-ref">Mapx output file description</span></a>.</p>
</div>
</div>
<div class="section" id="rtg-species-frequency-estimation">
<h2>RTG species frequency estimation<a class="headerlink" href="#rtg-species-frequency-estimation" title="Permalink to this headline">¶</a></h2>
<p>Use the following set of tasks to estimate the frequency of bacterial
species in a metagenomic sample. The RTG <code class="docutils literal"><span class="pre">species</span></code> frequency estimator,
called species, takes a set of reads mapped against a bacterial database
and from this estimates the relative frequency of each species in the
database.</p>
<p>Table : Overview of species frequency estimation tasks</p>
<table border="1" class="docutils">
<colgroup>
<col width="4%" />
<col width="25%" />
<col width="10%" />
<col width="61%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head" colspan="2">Task</th>
<th class="head">Command &amp; Utilities</th>
<th class="head">Purpose</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>1</td>
<td>Format reference data</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">format</span></code></td>
<td>Convert reference sequence from FASTA file to RTG Sequence Data Format (SDF)</td>
</tr>
<tr class="row-odd"><td>2</td>
<td>Format read data</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">format</span></code></td>
<td>Convert read sequence from FASTA and FASTQ files to RTG Sequence Data Format (SDF)</td>
</tr>
<tr class="row-even"><td>3</td>
<td>Run contamination filter (optional)</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">mapf</span></code></td>
<td>Produce the SDF file of reads which map to the contaminant and the SDF file of those that do not</td>
</tr>
<tr class="row-odd"><td>4</td>
<td>Map metagenomic reads against bacterial database</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">map</span></code></td>
<td>Generate read alignments against a given reference, and report in a BAM file for downstream analysis</td>
</tr>
<tr class="row-even"><td>5</td>
<td>Run species estimator</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">species</span></code></td>
<td>Produce a text file which contains a list of species, one per line, with an estimate of the relative frequency in the sample</td>
</tr>
</tbody>
</table>
<div class="section" id="task-1-format-reference-data-species">
<h3>Task 1 - Format reference data (species)<a class="headerlink" href="#task-1-format-reference-data-species" title="Permalink to this headline">¶</a></h3>
<p>RTG tools require a conversion of reference sequences from FASTA files
into the RTG SDF format. This task will be completed with the <code class="docutils literal"><span class="pre">format</span></code>
command. The conversion will create an SDF directory containing the
reference sequences.</p>
<p>Use the <code class="docutils literal"><span class="pre">format</span></code> command to convert multiple input files into a single
SDF directory for the reference database.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg format -o bacteria-db /data/bacteria/db/*.fa.gz
</pre></div>
</div>
<p>This takes the reference FASTA files and creates a directory called
<code class="docutils literal"><span class="pre">bacteria-db</span></code> containing the SDF. You can use the <code class="docutils literal"><span class="pre">sdfstats</span></code> command to
show statistics for your reference SDF.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg sdfstats bacteria-db
Type               : DNA
Number of sequences: 311276
Maximum length     : 13033779
Minimum length     : 0
Sequence names     : yes
N                  : 33864547
A                  : 4167856151
C                  : 4080877385
G                  : 4072353906
T                  : 4177108579
Total residues     : 16532060568
Residue qualities  : no
</pre></div>
</div>
<p>Alternatively a species reference SDF for running the <code class="docutils literal"><span class="pre">species</span></code> command
can be obtained from our website (<a class="reference external" href="http://www.realtimegenomics.com">http://www.realtimegenomics.com</a>).</p>
</div>
<div class="section" id="task-2-format-read-data-species">
<h3>Task 2 - Format read data (species)<a class="headerlink" href="#task-2-format-read-data-species" title="Permalink to this headline">¶</a></h3>
<p>RTG tools require a conversion of read sequence data from FASTA or FASTQ
files into the RTG SDF format. This task will be completed with the
<code class="docutils literal"><span class="pre">format</span></code> command. The conversion will create an SDF directory for the
sample reads.</p>
<p>Take a paired set of reads in FASTQ format and convert it into RTG data
format (SDF). This example shows one run of data, taking as input both
left and right mate pairs from the same run.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg format -f fastq -q sanger -o bacteria-sample \
  -l /data/reads/bacteria/sample_1.fq \
  -r /data/reads/bacteria/sample_2.fq
</pre></div>
</div>
<p>This creates a directory named <code class="docutils literal"><span class="pre">bacteria-sample</span></code> with two
subdirectories, named &#8216;left&#8217; and &#8216;right&#8217;. Use the <code class="docutils literal"><span class="pre">sdfstats</span></code> command to
verify this step.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg sdfstats bacteria-sample
</pre></div>
</div>
</div>
<div class="section" id="task-3-run-contamination-filter-optional">
<h3>Task 3 - Run contamination filter (optional)<a class="headerlink" href="#task-3-run-contamination-filter-optional" title="Permalink to this headline">¶</a></h3>
<p>Optionally filter the metagenomic read sample to remove human
contamination using Tasks 1 through 4 of <a class="reference internal" href="#rtg-contaminant-filtering"><span class="std std-ref">RTG contaminant filtering</span></a></p>
</div>
<div class="section" id="task-4-map-metagenomic-reads-against-bacterial-database">
<h3>Task 4 - Map metagenomic reads against bacterial database<a class="headerlink" href="#task-4-map-metagenomic-reads-against-bacterial-database" title="Permalink to this headline">¶</a></h3>
<p>Map the metagenomic reads against the reference database to generate
alignments in the BAM format (Binary Sequence Alignment/Map file
format). The read set in this example is paired end.</p>
<p>It is recommended that during mapping either the <code class="docutils literal"><span class="pre">--max-top-results</span></code>
flag be set to a high value, such as 100, or that the <code class="docutils literal"><span class="pre">--all-hits</span></code>
option be used. This helps ensure that all relevant species in the
database are accurately represented in the output. However, note that a
very large <code class="docutils literal"><span class="pre">--max-top-results</span></code> requires additional memory during
mapping.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg map -i bacteria-sample -t bacteria-db -o map-sample -n 100
</pre></div>
</div>
</div>
<div class="section" id="task-5-run-species-estimator">
<h3>Task 5 - Run species estimator<a class="headerlink" href="#task-5-run-species-estimator" title="Permalink to this headline">¶</a></h3>
<p>The species estimator, <code class="docutils literal"><span class="pre">species</span></code>, takes as input the BAM format files
from the mapping performed against the reference database.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg species -t bacteria-db -o species-result map-sample/alignments.bam
</pre></div>
</div>
<p>This run generates a new output directory <code class="docutils literal"><span class="pre">species_result</span></code>. The main
result file in this directory will be called <code class="docutils literal"><span class="pre">species.tsv</span></code>. In the
output the bacterial species are ordered from most to least abundant.
The output file can be directly loaded into a spreadsheet program like
Microsoft Excel.</p>
<p>The <code class="docutils literal"><span class="pre">species.tsv</span></code> file contains results for both species with associated
genomic sequences and internal nodes in the taxonomy. In some scenarios
it will only be necessary to examine those rows corresponding to
sequences in the database, such rows have a <code class="docutils literal"><span class="pre">Y</span></code> in the
<code class="docutils literal"><span class="pre">has-sequence</span></code> column. Internal taxonomy nodes (i.e. ones that have no
associated sequence data) always have a breadth and depth of coverage of
zero because no reads directly map to them. For further detail on the
<code class="docutils literal"><span class="pre">species.tsv</span></code> file format see <a class="reference internal" href="appendix.html#species-results-file-description"><span class="std std-ref">Species results file description</span></a></p>
<p>Also produced is an HTML5 summary file called <code class="docutils literal"><span class="pre">index.html</span></code> which
contains an interactive pie chart detailing the results.</p>
<p>The best results are obtained when as many relevant records as possible
are given to the species estimator. If you have insufficient memory to
use all your mapping results then using the filtering options may help.
You could filter the results by selecting mappings with good alignment
scores or mated only reads.</p>
</div>
</div>
<div class="section" id="rtg-sample-similarity">
<h2>RTG sample similarity<a class="headerlink" href="#rtg-sample-similarity" title="Permalink to this headline">¶</a></h2>
<p>Use the following set of tasks to produce a similarity matrix from the
comparison of a group of read sets. An example use case is in
metagenomics where several bacteria samples taken from different sites
need to be compared.</p>
<p>The <code class="docutils literal"><span class="pre">similarity</span></code> command performs a similarity analysis on multiple read
sets independent of any reference genome. It does this by examining
<em>k</em>-mer word frequencies and the intersections between sets of reads.</p>
<p>Table : Overview of sample similarity tasks</p>
<table border="1" class="docutils">
<colgroup>
<col width="5%" />
<col width="22%" />
<col width="16%" />
<col width="58%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Task</th>
<th class="head">&nbsp;</th>
<th class="head">Command &amp; Utilities</th>
<th class="head">Purpose</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>1</td>
<td>Prepare read sets</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">format</span></code></td>
<td>Convert reference sequence from FASTA file to RTG Sequence Data Format (SDF)</td>
</tr>
<tr class="row-odd"><td>2</td>
<td>Generate read set name map</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">text-editor</span></code></td>
<td>Produce the map of names to read set SDF locations</td>
</tr>
<tr class="row-even"><td>3</td>
<td>Run similarity tool</td>
<td><code class="docutils literal"><span class="pre">$</span> <span class="pre">rtg</span> <span class="pre">similarity</span></code></td>
<td>Process the read sets for similarity</td>
</tr>
</tbody>
</table>
<div class="section" id="task-1-prepare-read-sets">
<h3>Task 1 - Prepare read sets<a class="headerlink" href="#task-1-prepare-read-sets" title="Permalink to this headline">¶</a></h3>
<p>RTG tools require a conversion of read sequence data from FASTA or FASTQ
files into the RTG SDF format. This task will be completed with the
format command. The conversion will create an SDF directory for the
sample reads.</p>
<p>Take a paired set of reads in FASTQ format and convert it into RTG data
format (SDF). This example shows one run of data, taking as input both
left and right mate pairs from the same run.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg format -f fastq -q sanger -o /data/reads/read-sample1-sdf \
  -l /data/reads/fastq/read-sample1_1.fq \
  -r /data/reads/fastq/read-sample2_2.fq
</pre></div>
</div>
<p>This creates a directory named &#8216;read-sample1-sdf&#8217; with two subdirectories,
named &#8216;left&#8217; and &#8216;right&#8217;. Use the <code class="docutils literal"><span class="pre">sdfstats</span></code> command to verify this step.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg sdfstats /data/reads/read-sample1-sdf
</pre></div>
</div>
<p>Repeat for all read samples to be compared. This example shows how this
can be done with the format command in a loop.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ for left_fq in /data/reads/fastq/*_1.fq; do
$ right_fq=${left_fq/_1.fq/_2.fq}
$ sample_id=$(basename ${left_fq/_1.fq})
$ rtg format -f fastq -q sanger -o /data/reads/${sample_id}-sdf -l ${left_fq} \
  -r ${right_fq}
$ done
</pre></div>
</div>
</div>
<div class="section" id="task-2-generate-read-set-name-map">
<h3>Task 2 - Generate read set name map<a class="headerlink" href="#task-2-generate-read-set-name-map" title="Permalink to this headline">¶</a></h3>
<p>With a text editor, or other tool, create a text file containing a list
of sample name to sample read SDF file locations. If two or more read
sets are from the same sample they can be combined by giving them the
same sample name in the file list.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ cat read-set-list.txt
sample1 /data/reads/read-sample1-sdf
sample2 /data/reads/read-sample2-sdf
sample3 /data/reads/read-sample3-sdf
sample4 /data/reads/read-sample4-sdf
sample5 /data/reads/read-sample5-sdf
</pre></div>
</div>
</div>
<div class="section" id="task-3-run-similarity-tool">
<h3>Task 3 - Run similarity tool<a class="headerlink" href="#task-3-run-similarity-tool" title="Permalink to this headline">¶</a></h3>
<p>Run the <code class="docutils literal"><span class="pre">similarity</span></code> command setting the <em>k</em>-mer word size (<code class="docutils literal"><span class="pre">-w</span></code>
parameter) and the step size (<code class="docutils literal"><span class="pre">-s</span></code> parameter) on the read sets by
specifying the file listing the read sets. Some experimentation should
be performed with different word and step size parameters to find good
trade-offs between memory usage and run time. Should it be necessary to
reduce the memory used it is possible to limit the number of reads used
from each SDF by specifying the <code class="docutils literal"><span class="pre">--max-reads</span></code> parameter.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ rtg similarity -w 25 -s 25 --max-reads 1000000 -I read-set-list.txt \
  -o similarity-output
</pre></div>
</div>
<p>The program puts its output in the specified output directory.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ ls similarity-output/
 4693 Aug 29 20:17 closest.tre
19393 Aug 29 20:17 closest.xml
   33 Aug 29 20:17 done
11363 Aug 29 20:17 similarity.log
48901 Aug 29 20:17 similarity.tsv
  693 Aug 29 20:17 progress
</pre></div>
</div>
<p>The <code class="docutils literal"><span class="pre">similarity.tsv</span></code> file is a tab separated file containing a matrix of
counts of the number of <em>k</em>-mers in common between each pair of samples.
The <code class="docutils literal"><span class="pre">closest.tre</span></code> and <code class="docutils literal"><span class="pre">closest.xml</span></code> files are nearest neighbor trees
built from the counts from the similarity matrix. The <code class="docutils literal"><span class="pre">closest.tre</span></code> is
in Newick format and the <code class="docutils literal"><span class="pre">closest.xml</span></code> file is phyloXML. The
<code class="docutils literal"><span class="pre">similarity.pca</span></code> file contains a principal component analysis on the
similarity matrix in <code class="docutils literal"><span class="pre">similarity.tsv</span></code>.</p>
<p>You may wish to view <code class="docutils literal"><span class="pre">closest.tre</span></code> or <code class="docutils literal"><span class="pre">closest.xml</span></code> in your preferred
tree viewing tool or use the principal component analysis output in
<code class="docutils literal"><span class="pre">similarity.pca</span></code> to produce a three-dimensional grouping plot showing
visually the clustering of samples.</p>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="administration.html" title="Administration &amp; Capacity Planning"
             >next</a> |</li>
        <li class="right" >
          <a href="rtg_command_reference.html" title="RTG Command Reference"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">RTG Core Operations Manual v3.7</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2016, Real Time Genomics.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.6.
    </div>
  </body>
</html>